---
title: "Fungi 2012 project - Differential expression analysis for KEGG ortholog network"
author: "Alonso Serrano"
date: "`r Sys.Date()`"
output:
 html_document:
   toc: true
   number_sections: true
---

# Introduction
This file contains the script and results for the differential expression analysis done to the
fungi 2012 data, based on the KEGG orthologs.

*20191121 Creation  
*20200110 Updated with 2020-01-08 files from John  
*20200401 Excluding 5 year fertilisation data  
*20200407 Adding Fert vs Control DE  

## Prerequisites
KEGG ortholog file

A metadata file containing all the information related to each sample, file name, time, condition and replica number.

# Setup
Load libraries
```{r, message=FALSE}
library(DESeq2)
library(tximport)
library(BiocParallel)
library(plotly)
library(dplyr)
library(here)
register(MulticoreParam(4))
source(here("Rtoolbox/src/utilsDE.r"))
source(here("/Rtoolbox/src/plot3Dvector.R"))
source(here("/Rtoolbox/src/plotVectorPCA.R"))
source(here("Rtoolbox/src/plotEnrichedTreemap.R"))
source(here("Rtoolbox/src/plotUpAndDown.R"))
source(here("UPSCb-common/src/R/gopher.R"))

```
```{r}
plotResults <- function(theData) {
  
  for (deName in names(theData)) {
  
    #deName <- paste0('Cluster',clus)
    #clusIndex <- match(deName, names(theClusters))
    #clusIndex <- as.numeric(substr(deName,8,nchar(deName)))
    cat(paste0("<H3>", deName, "","</H3>"))
    cat("\n\n")
    
    enrUp <- theData[[deName]]$up[['ko_pathway']]
    enrDown <- theData[[deName]]$down[['ko_pathway']]
    
    cat("Up genes \n\n")
    if (!is.null(enrUp) ) {
      plotEnrichedTreemap(theData[[deName]]$up, enrichment = 'ko_pathway', namespace = 'none',
                          de='up', title = paste(deName," KO enrichment"))
    } else {
      cat("No enrichment data found")
    }
    cat("\n")
    cat("<hr>")
    
    cat("Down genes \n\n")

    
    if (!is.null(enrDown) ) {
      plotEnrichedTreemap(theData[[deName]]$down, enrichment = 'ko_pathway', namespace = 'none',
                          de='down', title = paste(deName," KO enrichment"))
    } else {
      cat("No enrichment data found")
    }
    cat("\n")
    cat("<hr>")

  }
  
}

getDEenrichment <- function(deData, background){
  results <- lapply(deData, function(x){

    upGenes <- rownames(x[x$log2FoldChange>0,])
    downGenes <- rownames(x[x$log2FoldChange<0,])
    upEnr <- list(ko_pathway=NULL)
    downEnr <- list(ko_pathway=NULL)
    
    if(length(upGenes)>1) 
      upEnr <- gopher(upGenes, task=list("ko_pathway"), background=background, url='ko', alpha=0.05)

    if(length(downGenes)>1) 
      downEnr <- gopher(downGenes, task=list("ko_pathway"), background=background, url='ko', alpha=0.05)
    
    res <- list(up=upEnr, down=downEnr)
    
    res
  })
  
  results
}
```


Set the project folder

```{r}
projectFolder <- "/mnt/picea/projects/spruce/vhurry/fungi-flakaliden-2012"
deFolder <- file.path(projectFolder,"DE")
dataFolder <- file.path(projectFolder,"data")
```
```{r carga, echo=FALSE, eval=F}
load(file.path(deFolder,"allDEresults.RData"))
```

Read metadata

```{r}
meta <- read.table(file.path(dataFolder, "meta.tsv"), header = T,
                        sep=';', comment.char = "", stringsAsFactors = F)
```


Add the interaction group to the metadata, in order to simplify
```{r}
head(meta, n=3)
meta$date <- factor(meta$date, levels = c("Early_June", "Late_June", "August", "October") )

```
We remove the 5 year fertilization samples as we want to compare only control vs 25 year fertilization
```{r remove5year}
meta$treatment <- ifelse(meta$treatment=="25_year", "Fertilised", meta$treatment)
meta <- meta[meta$treatment != "5_year",]
meta$treatment <- factor(meta$treatment, levels = c("Control", "Fertilised") )
```
```{r}
meta$group <- paste0(meta$treatment, "_", meta$date)
meta$group <- factor(meta$group)
```

Load counts data
```{r}

data <- read.table(file.path(dataFolder, "kos.raw.tsv"), header = T,
                       sep='\t', comment.char = "", quote="",stringsAsFactors = F, row.names = 1)
data <- data[!rownames(data)=='Unclassified',]
rownames(data) <- substr(rownames(data), 1, 6)

# remove non usefull samples
data <- data[,colnames(data) %in% meta$SciLifeID]

#' #Order them by group
data <- data[, match(meta$SciLifeID, colnames(data))]

```

Import data as a DESeq object and place the sample names to the columns
```{r deImport, message=FALSE}
ddsMatrix <- DESeqDataSetFromMatrix(data, colData=meta, design = ~group)
```
check the DESeqDataSet
```{r}
ddsMatrix
```

Some KOs have 0 expression accross all samples we can remove it
```{r}
ddsMatrix <- ddsMatrix[(rowSums(counts(ddsMatrix)) > 0),]
ddsMatrix
```


# QA
## PCA
We use blind vst to plot a PCA and check for patterns and outliers

```{r}
vsd.QA <- vst(ddsMatrix, blind=TRUE)
geneNumber <- length(rownames(assay(vsd.QA)))
```
PCA using only treatment metadata
```{r}
plotPCA(vsd.QA, intgroup = c("treatment"), ntop=geneNumber)
```

PCA using only date metadata
```{r}
plotPCA(vsd.QA, intgroup = c("date"), ntop=geneNumber)
```
The PCA seems correct. We plot the same PCAs but using different colors to focus on the time or the treatment

PCA using both treatment and date
```{r}
plotPCA(vsd.QA, intgroup = c("treatment", "date"), ntop=geneNumber)
```

The PCAs seems correct. 
```{r}
save(vsd.QA, meta, ddsMatrix, file = file.path(projectFolder, "DE", "vsd_QA.RData"))
```

## Focused PCA
We plot a PCA with letters instead of numbers for increased clarity
```{r}

vsd.QA.t <- t(assay(vsd.QA))
pca <- prcomp(vsd.QA.t)
percents <- round(summary(pca)$importance[2,]*100)
pchs <- LETTERS[as.integer(factor(meta$date))]
cols <- c("#fc8d62", "#66c2a5")[as.integer(factor(meta$treatment))]
PCAlegend <- data.frame(pchs, meta$date)

par(oma = c(4, 1, 1, 4))
plot(pca$x[,1], pca$x[,2],
     xlab=paste("Comp. 1 (",percents[1],"%)",sep=""),
     ylab=paste("Comp. 2 (",percents[2],"%)",sep=""),
     col=cols,pch=pchs, cex = 1.0, main="")
par(fig = c(0, 1, 0, 1), oma = c(0, 0, 0, 0), mar = c(0, 0, 0, 0), new = TRUE)
plot(0, 0, type = "n", bty = "n", xaxt = "n", yaxt = "n")
legend("bottomleft", legend=unique(PCAlegend$pchs:meta$date), cex = 0.8, xpd=TRUE)
legend("bottomright", legend=unique(meta$treatment), fill=unique(cols),cex = 0.8, xpd=TRUE)

```


## Scree plot
We can see how PC proportions are distributed
```{r}
screePlot(pca, 10)
```



The first 5 components explain 55 % of the variance
```{r}
summary(pca)$importance[2,][1:5] *100
sum(summary(pca)$importance[2,][1:5]) * 100
```



## PCA bi-plot comparition

As seen before, we can focus on the 5 components and plot a comparative bi-plot
```{r}
par(cex=1.0, cex.axis=0.8, cex.main=0.8)
pairs(pca$x[,1:5], col=cols, main="Principal components analysis bi-plot\nPCs 1-5", pch=pchs)
```



## 3D PCA
We represent the first 3 compoments as a 3D interactive object

```{r}
colors <- c("#f8766d", "#7cae00","#00bfc4")

df3D <- data.frame(treatment = meta$treatment)
df3D$time <- meta$date
df3D$colors <- colors[as.numeric(df3D$treatment)]
df3D$PC1 <- pca$x[,1]
df3D$PC2 <- pca$x[,2]
df3D$PC3 <- pca$x[,3] 

plot_ly(df3D, x = ~PC1, y = ~PC3, z = ~PC2, 
        color = ~treatment, colors = colors,
        text = ~paste("Date:", time, '<br>Treatment:', treatment) ) %>%   
  add_markers() %>%
  layout(scene = list(xaxis = list(title = paste("PC1 (",percents[1],"%)",sep="")),
                      yaxis = list(title = paste("PC3 (",percents[3],"%)",sep="")),
                      zaxis = list(title = paste("PC2 (",percents[2],"%)",sep=""))))

```



# Differential Expression
The dataset is ready for DESeq2. We check the design of the DE object 

```{r}
design(ddsMatrix)
```

We are interested on future comparitions like DE ko in each timepoints or between timepoints
that is why we created the interaction group

## Run DESeq
```{r runDE, eval=T}
dds <- DESeq(ddsMatrix)
```

We check the result names
```{r}
resultsNames(dds)
```

In order to make all the comparitions automated, we prepare vectors with the elements that will be used
in the result extraction
```{r}
controlVector <- as.character(unique(meta$group[grep("Control", meta$group)]))
fertilisedVector <- as.character(unique(meta$group[grep("Fertilised", meta$group)]))
```

## Result extraction
We want to make 3 different comparations, each one composed of several outputs from DESeq2
In order to do that, we will extract all the results for each comparation type, store them
as a list and then we can process them.

### Control date vs date
For this comparition we take consider only the control condition and we compare each pair of weeks
to determine which genes are up or down regaluated from one week to the previous week

```{r eval=T}
control.res <- mapply(getRes, controlVector[-1],
                      controlVector[-length(controlVector)], 
                      MoreArgs = list(localDDS=dds, group="group"))
```

### Fertilised  date vs date
For this comparition we take consider only the fertilised condition and we compare each pair of weeks
to determine which genes are up or down regaluated from one week to the previous week

```{r eval=T}
fertilised.res <- mapply(getRes, fertilisedVector[-1],
                          fertilisedVector[-length(fertilisedVector)], 
                          MoreArgs = list(localDDS=dds, group="group"))
```



We add the corresponding names to the results
```{r eval=T}
names(control.res) <- c("Control_Late_JunevsControl_Early_June", 
                        "Control_AugustvsControl_Late_June", 
                        "Control_OctobervsControl_August")
names(fertilised.res) <- c("Fertilised_Late_JunevsFertilised_Early_June", 
                        "Fertilised_AugustvsFertilised_Late_June", 
                        "Fertilised_OctobervsFertilised_August")

```

### Timepoint Fertilised  vs Control
For this comparition we take consider only a date each time and then compare DE genes due to 25 year fertilisation

```{r eval=T}
combined.res <- mapply(getRes, fertilisedVector, controlVector,
                            MoreArgs = list(localDDS=dds, group="group"))
```

We add the corresponding names to the results
```{r eval=T}

names(combined.res) <- c("Fertilised_Early_JunevsControl_Early_June", 
                              "Fertilised_Late_JunevsControl_Late_June", 
                              "Fertilised_AugustvsControl_August", 
                              "Fertilised_OctobervsControl_October")
```
```{r include=FALSE, eval=T}
save(dds, control.res, fertilised.res, combined.res, file=file.path(deFolder,"allDEresults.RData"))
```

```{r include=FALSE, eval=F}
load(file.path(deFolder,"allDEresults.RData"))
```

## Filtering
The results contain all the genes, we need to filter them to get only those genes
that are stadistical significant and that have a strong condtition effect
For that we discard all the genes that has a padj higher than 0.01 and
a log2FoldChange lower than 0.5 (absolute value)
```{r filtering, results="hide", message=FALSE}
control.res.filter <- lapply(control.res, filterDE)
fertilised.res.filter <- lapply(fertilised.res, filterDE)
combined.res.filter <- lapply(combined.res, filterDE)
```

It is a good practice to do a check and see if the filtering was correct
```{r checkfilter, results="hide", message=FALSE}
lapply(control.res.filter, function(x) print(all(abs(x$log2FoldChange)>0.5 & x$padj<0.01)) )
lapply(fertilised.res.filter, function(x) print(all(abs(x$log2FoldChange)>0.5 & x$padj<0.01)) )
lapply(combined.res.filter, function(x) print(all(abs(x$log2FoldChange)>0.5 & x$padj<0.01)) )
```

## Plot DE genes
We can plot the results as positive and negative bars to reflect the amount of genes up or down
regulated in each comparition 
For that we create a couple more vector to present the results



### Control week+1 vs week plot
```{r}
plotUpAndDown(control.res.filter)
```

### Fertilised week+1 vs week plot
```{r}
plotUpAndDown(fertilised.res.filter)
```

### Fertilised vs Control per week plot
```{r}
plotUpAndDown(combined.res.filter)
```

## DE Fertilisation vs Control
```{r}
design(ddsMatrix) <- ~treatment

ddsFungiTreatment  <- DESeq(ddsMatrix)

resFungiTreatment <- list( FertilisedvsControl = results(ddsFungiTreatment ,
        name= "treatment_Fertilised_vs_Control",
        filter = rowMedians(counts(ddsFungiTreatment )),
        parallel = FALSE) 
        )

save(ddsFungiTreatment, resFungiTreatment, file=file.path(projectFolder, "DE", "ddsTreatment.RData"))

resFungiTreatment.filter <- lapply(resFungiTreatment, filterDE)

plotUpAndDown(resFungiTreatment.filter)

```

# Results export
## Background for enrichment
We need to extract the names of all the genes that at least have 1 count across all samples to establish a background 
needed for future enrichment steps
```{r}
write.table(colnames(t(assay(dds))), file=file.path(deFolder,"fungiKO2012background.txt"), row.names = F,col.names = F, quote = F )
```

## Data for networks
For a co-expression network construction we need the vst data after doing a zeroMAD (Mean Decrease Accuracy) cleanup.
That will remove genes with zero or almost zero expression.
For the network construction with Seidr we need a headless files with the data and a file only with the gene names

### Combined VST data
```{r}
vsdCombined <- vst(dds, blind=FALSE)
combinedData <- t(assay(vsdCombined))
combinedData <- combinedData[,which(colMads(as.matrix(combinedData)) > 0)]
write.table(combinedData, file =file.path(deFolder, "fungiKO2012Combined.tsv"), 
            sep='\t', row.names=FALSE, col.names=FALSE)
writeLines(paste(colnames(combinedData), collapse = '\t'),
           con=file.path(deFolder,"fungiKO2012CombinedNames.txt"))
```


## DE results
For downstream analysis we export each set of results to a tsv file containing the padj and log2foldchange values for each comparition.
We export the results unfiltered
```{r exportDEresults, results="hide", message=FALSE}
de4network(control.res, file.path(deFolder,"deControl.tsv"))
de4network(fertilised.res, file.path(deFolder,"deFertilised.tsv"))
de4network(combined.res, file.path(deFolder,"deCombined.tsv"))

```

# Session information
```{r}
sessionInfo()
```




# References
