---
title: "Species and genus analysis."
author: "Alonso Serrano"
date: "2020/11/24"
output:
 html_document:
   toc: true
   number_sections: true
---

<style>
* {
  box-sizing: border-box;
}

/* Create two equal columns that floats next to each other */
.column {
  float: left;
  width: 50%;
  padding: 10px;
  
}

/* Clear floats after the columns */
.row:after {
  content: "";
  display: table;
  clear: both;
}
</style>

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```



# Introduction
This analysis focuses on the study of the enrichment tests and KO proportions of 
Piloderma genus, Cortinarius genus, Piloderma croceum and Cortinarius glaucopus.

## Prerrequisites:
- VST data of the fungi transcriptome dataset 
- Annotations file of the fungi transcriptome  
- Transcriptome taxonomy file of the fungi transcriptome  


# Setup
Load libraries
```{r}
library(here)
source(here("Rtoolbox/src/getEigengenes.R"))
source(here("Rtoolbox/src/plotEnrichedTreemap.R"))
source(here("/UPSCb-common/src/R/gopher.R"))
source("~/Git/rBiobox/R/plotEigengene.R")
source("~/scripts/koPathwayDiseaesCleanup/koProcessing.R")

```

Load data
```{r}
# fungi data holder
load("/mnt/picea/projects/spruce/vhurry/fungi-flakaliden-2011/data/megaData.RData")
# whole fungi metadata file
megaMeta <- read.table("/mnt/picea/projects/spruce/vhurry/fungi-flakaliden-2011/data/annotation_results.emapper.annotations", header = F, sep='\t', quote = "", comment.char = "")
# vst data for control and fertilised
load("/mnt/picea/projects/spruce/vhurry/fungi-flakaliden-2011/DE/transcriptsControlVsdData.RData")
load("/mnt/picea/projects/spruce/vhurry/fungi-flakaliden-2011/DE/transcriptsFertilisedVsdData.RData")
```

Extract metada for the plots
```{r}
weeksC <- substr(rownames(controlData), 2, 3)
treatmentC <- substr(rownames(controlData), 4, 4)
weeksF <- substr(rownames(fertilisedData), 2, 3)
treatmentF <- substr(rownames(fertilisedData), 4, 4)
```

Extract data for both genus
```{r}
cortinariusData <- megaData[megaData$genus == "Cortinarius",]
pilodermaData <- megaData[megaData$genus == "Piloderma",]
```

Apply DE filtering, abs(log2fc) > 0.5 and adjusted p-value < 0.01
```{r}
cortinariusData <- cortinariusData[(!is.na(cortinariusData$treatment_padj) & 
  cortinariusData$treatment_padj < 0.01 &
  abs(cortinariusData$treatment_lfc) > 0.5 ),]

pilodermaData <- pilodermaData[(!is.na(pilodermaData$treatment_padj) & 
  pilodermaData$treatment_padj < 0.01 &
  abs(pilodermaData$treatment_lfc)  > 0.5 ),]
```

Extract data for both species and unclassified categories with the genus
```{r}
data_cortinarius_glaucopus <- cortinariusData[cortinariusData$species =="Cortinarius glaucopus", ]
data_unclassified_cortinarius <- cortinariusData[cortinariusData$species =="Unclassified.Cortinarius", ]

data_piloderma_croceum <- pilodermaData[pilodermaData$species =="Piloderma croceum", ]
data_unclassified_piloderma <- pilodermaData[pilodermaData$species =="Unclassified.Piloderma", ]

```

Extract gene names
```{r}
genes_cortinarius_glaucopus <- data_cortinarius_glaucopus$gene
genes_unclassified_cortinarius <- data_unclassified_cortinarius$gene
genes_genus_cortinarius <- cortinariusData$gene

genes_piloderma_croceum <- data_piloderma_croceum$gene
genes_unclassified_piloderma <- data_unclassified_piloderma$gene
genes_genus_piloderma <- pilodermaData$gene
```

Enrichment tests for cortinarius
```{r eval=FALSE, include=T}
# species with genus background
cortinarius_glaucopus_enr <- gopher(genes_cortinarius_glaucopus, 
                                    background = genes_genus_cortinarius,
                                    url=paste0('fungi2011'), 
                                    task=c("cog","ko", "kog", "ko_pathway")
                                    )
# unclassified with genus background
unclassified_cortinarius_enr <- gopher(genes_unclassified_cortinarius, 
                                      background = genes_genus_cortinarius,
                                      url=paste0('fungi2011'), 
                                      task=c("cog","ko", "kog", "ko_pathway")
                                      )
# genus with whole background
cortinarius_glaucopus_full_enr <- gopher(genes_cortinarius_glaucopus, 
                                    url=paste0('fungi2011'), 
                                    task=c("cog","ko", "kog", "ko_pathway")
                                    )

# unclassified with whole background
unclassified_cortinarius_full_enr <- gopher(genes_unclassified_cortinarius, 
                                      url=paste0('fungi2011'), 
                                      task=c("cog","ko", "kog", "ko_pathway")
                                      )
# species with whole background
genus_cortinarius_enr <- gopher(genes_genus_cortinarius, 
                                url=paste0('fungi2011'), 
                                task=c("cog","ko", "kog", "ko_pathway")
                                )

```

Enrichment tests for piloderma
```{r eval=FALSE, include=T}
# species with genus background
piloderma_croceum_enr <- gopher(genes_piloderma_croceum, 
                                    background = genes_genus_piloderma,
                                    url=paste0('fungi2011'), 
                                    task=c("cog","ko", "kog", "ko_pathway")
                                    )

# unclassified with genus background
unclassified_piloderma_enr <- gopher(genes_unclassified_piloderma, 
                                      background = genes_genus_piloderma,
                                      url=paste0('fungi2011'), 
                                      task=c("cog","ko", "kog", "ko_pathway")
                                      )
# genus with whole background
piloderma_croceum_full_enr <- gopher(genes_piloderma_croceum, 
                                    url=paste0('fungi2011'), 
                                    task=c("cog","ko", "kog", "ko_pathway")
                                    )

# unclassified with whole background
unclassified_piloderma_full_enr <- gopher(genes_unclassified_piloderma,
                                      url=paste0('fungi2011'), 
                                      task=c("cog","ko", "kog", "ko_pathway")
                                      )
# species with whole background
genus_piloderma_enr <- gopher(genes_genus_piloderma, 
                                url=paste0('fungi2011'), 
                                task=c("cog","ko", "kog", "ko_pathway")
                                )

```

```{r eval=FALSE, include=FALSE}
#save intermediate results for fast processing
save(unclassified_cortinarius_enr,cortinarius_glaucopus_enr,
     cortinarius_glaucopus_full_enr, unclassified_cortinarius_full_enr,
     genus_cortinarius_enr, 
     piloderma_croceum_enr, unclassified_piloderma_enr,
     piloderma_croceum_full_enr, unclassified_piloderma_full_enr,
     genus_piloderma_enr,
     file="tempEnr_201125.RData")
```

```{r include=FALSE}
# load intermediate results
load("tempEnr_201125.RData")
```

# Enrichment tests
## Cortinarius
```{r , message=FALSE, warning=FALSE, results='asis'}
listOfThings <- list(cortinarius_glaucopus_enr, unclassified_cortinarius_enr,
                  cortinarius_glaucopus_full_enr, unclassified_cortinarius_full_enr,
                  genus_cortinarius_enr)

names(listOfThings) <- c("Cortinarius glaucopus</br>Background cortinarius genus</br>", 
                         "Unclassified cortinarius</br>Background cortinarius genus</br>",
                         "Cortinarius glaucopus</br>Background full dataset</br>",
                         "Unclassified cortinarius</br>Background full dataset</br>",
                         "Genus Cortinarius</br>Background full dataset</br>")

for(currentEnrIndex in 1:length(listOfThings)){
  currentEnr <- listOfThings[[currentEnrIndex]]
  
  cat(paste0("<H3>", names(listOfThings)[currentEnrIndex], "","</H3>"))
  cat("\n\n")
    
  if(!is.null(currentEnr$ko_pathway)) {
    currentEnr$ko_pathway <- koPathwayDiseaseCleanup(koTranslate(currentEnr$ko_pathway))
    
    plotEnrichedTreemap(currentEnr, "ko_pathway", "none", title = "KO PATHWAY")
    cat("\n\n")
  }
  
  if(!is.null(currentEnr$cog))
    plotEnrichedTreemap(currentEnr, "cog", "none", title = "COG")
  cat("\n\n")  
  
  if(!is.null(currentEnr$ko)) {
    currentEnr$ko <- koTranslate(currentEnr$ko)
    plotEnrichedTreemap(currentEnr, "ko", "none", title = "KO")
  }
  cat("\n\n")
  
  if(!is.null(currentEnr$kog))
    plotEnrichedTreemap(currentEnr, "kog", "none", title = "KOG")
  cat("\n\n")
}
cat("\n\n")
cat("\n\n")

```

## Piloderma
```{r , message=FALSE, warning=FALSE, results='asis'}
listOfThings <- list(piloderma_croceum_enr, unclassified_piloderma_enr,
                  piloderma_croceum_full_enr, unclassified_piloderma_full_enr,
                  genus_piloderma_enr)

names(listOfThings) <- c("piloderma croceum</br>Background piloderma genus</br>", 
                         "Unclassified piloderma</br>Background piloderma genus</br>",
                         "piloderma croceum</br>Background full dataset</br>",
                         "Unclassified piloderma</br>Background full dataset</br>",
                         "Genus piloderma</br>Background full dataset</br>")

for(currentEnrIndex in 1:length(listOfThings)){
  currentEnr <- listOfThings[[currentEnrIndex]]

  cat(paste0("<H3>", names(listOfThings)[currentEnrIndex], "","</H3>")) 
  cat("\n\n")

  if(!is.null(currentEnr$ko_pathway)) {
    currentEnr$ko_pathway <- koPathwayDiseaseCleanup(koTranslate(currentEnr$ko_pathway))
    
    plotEnrichedTreemap(currentEnr, "ko_pathway", "none", title = "KO PATHWAY")
    cat("\n\n")
  }
  
  if(!is.null(currentEnr$cog))
    plotEnrichedTreemap(currentEnr, "cog", "none", title = "COG")
  cat("\n\n")  
  
  if(!is.null(currentEnr$ko)) {
    currentEnr$ko <- koTranslate(currentEnr$ko)
    plotEnrichedTreemap(currentEnr, "ko", "none", title = "KO")
  }
  cat("\n\n")
  
  if(!is.null(currentEnr$kog))
    plotEnrichedTreemap(currentEnr, "kog", "none", title = "KOG")
  cat("\n\n")
}
cat("\n\n")
cat("\n\n")

```

# KO analysis
## Cortinarius
Unclassified cortinarius KO extraction
```{r}
kos_unclassified_cortinarius <- megaMeta[megaMeta$V1 %in% genes_unclassified_cortinarius,]$V9
kos_unclassified_cortinarius <- gsub("ko:", "", kos_unclassified_cortinarius)
kos_unclassified_cortinarius <- kos_unclassified_cortinarius[kos_unclassified_cortinarius != ""]

kos_unclassified_cortinarius <- unique(unlist(strsplit(kos_unclassified_cortinarius, ",")))

```

Cortinarius glaucopus KO extraction
```{r}
kos_cortinarius_glaucopus <- megaMeta[megaMeta$V1 %in% genes_cortinarius_glaucopus,]$V9
kos_cortinarius_glaucopus <- gsub("ko:", "", kos_cortinarius_glaucopus)
kos_cortinarius_glaucopus <- kos_cortinarius_glaucopus[kos_cortinarius_glaucopus != ""]

kos_cortinarius_glaucopus <- unique(unlist(strsplit(kos_cortinarius_glaucopus, ",")))
```

Exclusive and common KOs
```{r}
kos_cortinarius_genus <- union(kos_unclassified_cortinarius, kos_cortinarius_glaucopus)
exclusive_kos_unclassified_cortinarius <- setdiff(kos_unclassified_cortinarius,kos_cortinarius_glaucopus)
exclusive_kos_cortinarius_glaucopus <- setdiff(kos_cortinarius_glaucopus,kos_unclassified_cortinarius)
common_kos_cortinarius_species <- intersect(kos_cortinarius_glaucopus,kos_unclassified_cortinarius)
```

Cortinarius genus distribution
```{r}
print(paste0("Cortinarius glaucopus over cortinarius genus: ",
      round(length(exclusive_kos_cortinarius_glaucopus)/
              length(kos_cortinarius_genus)*100,2)))

print(paste0("Unknown cortinarius over cortinarius genus: ",
      round(length(exclusive_kos_unclassified_cortinarius)/
              length(kos_cortinarius_genus)*100,2)))

print(paste0("Common between species inside cortinarius genus: ",
      round(length(common_kos_cortinarius_species)/
              length(kos_cortinarius_genus)*100,2)))
```

## Piloderma
Unclassified piloderma KO extraction
```{r}
kos_unclassified_piloderma <- megaMeta[megaMeta$V1 %in% genes_unclassified_piloderma,]$V9
kos_unclassified_piloderma <- gsub("ko:", "", kos_unclassified_piloderma)
kos_unclassified_piloderma <- kos_unclassified_piloderma[kos_unclassified_piloderma != ""]

kos_unclassified_piloderma <- unique(unlist(strsplit(kos_unclassified_piloderma, ",")))

```

Piloderma croceum KO extraction
```{r}
kos_piloderma_croceum <- megaMeta[megaMeta$V1 %in% genes_piloderma_croceum,]$V9
kos_piloderma_croceum <- gsub("ko:", "", kos_piloderma_croceum)
kos_piloderma_croceum <- kos_piloderma_croceum[kos_piloderma_croceum != ""]

kos_piloderma_croceum <- unique(unlist(strsplit(kos_piloderma_croceum, ",")))
```

Exclusive and common KOs
```{r}
kos_piloderma_genus <- union(kos_unclassified_piloderma, kos_piloderma_croceum)
exclusive_kos_unclassified_piloderma <- setdiff(kos_unclassified_piloderma,kos_piloderma_croceum)
exclusive_kos_piloderma_croceum <- setdiff(kos_piloderma_croceum,kos_unclassified_piloderma)
common_kos_piloderma_species <- intersect(kos_piloderma_croceum,kos_unclassified_piloderma)
```

Piloderma genus distribution
```{r}
print(paste0("Piloderma croceum over piloderma genus: ",
      round(length(exclusive_kos_piloderma_croceum)/
              length(kos_piloderma_genus)*100,2)))

print(paste0("Unknown piloderma over piloderma genus: ",
      round(length(exclusive_kos_unclassified_piloderma)/
              length(kos_piloderma_genus)*100,2)))

print(paste0("Common between species inside piloderma genus: ",
      round(length(common_kos_piloderma_species)/
              length(kos_piloderma_genus)*100,2)))
```


# KO expression profile plots
We select 10 random common KOs betweem the species and the unclassified category, 
then we reverse look for the genes assigned to those KOs and plot them individually 
in both ND and NE conditions.

Cortinarius plot
```{r}
# get 10 common KOs between species
corti10 <- sample(common_kos_cortinarius_species,10)
print("Selected KOs: ")
print(corti10)
cortiGlaucopusData <- megaMeta[megaMeta$V1 %in% genes_cortinarius_glaucopus,]
cortiGlaucopusKOgenes <- unique(cortiGlaucopusData[grepl(paste(corti10,collapse="|"), cortiGlaucopusData$V9),]$V1)

unkCortiData <- megaMeta[megaMeta$V1 %in% genes_unclassified_cortinarius,]
unkCortiKOgenes <- unique(unkCortiData[grepl(paste(corti10,collapse="|"), unkCortiData$V9),]$V1)
```

CORTINARIUS GLAUCOPUS genes to KO profiles
```{r echo=FALSE, results='asis',message=FALSE, warning=FALSE,  fig.hold='hold', out.width="50%"}
p1 <- plotEigengene(controlData, intersect(cortiGlaucopusKOgenes,colnames(controlData)), weeksC, treatmentC)

p2 <- plotEigengene(fertilisedData, intersect(cortiGlaucopusKOgenes,colnames(fertilisedData)), weeksF, treatmentF, colors = "#BE9230")

cat("\n\n")
cat("\n\n")
par(mar = c(4, 4, 0.1, 0.1))
print(p1)
print(p2)
par(mfrow=c(1,1))
cat("\n\n")
cat("\n\n")
```
UNKNOWN CORTINARIUS genes to KO profiles
```{r echo=FALSE, results='asis',message=FALSE, warning=FALSE,  fig.hold='hold', out.width="50%"}
p1 <- plotEigengene(controlData, intersect(unkCortiKOgenes,colnames(controlData)), weeksC, treatmentC)

p2 <- plotEigengene(fertilisedData, intersect(unkCortiKOgenes,colnames(fertilisedData)), weeksF, treatmentF, colors = "#BE9230")

cat("\n\n")
cat("\n\n")
par(mar = c(4, 4, 0.1, 0.1))
print(p1)
print(p2)
par(mfrow=c(1,1))
cat("\n\n")
cat("\n\n")
```

Piloderma plot
```{r}
# get 10 common KOs between species
pilo10 <- sample(common_kos_piloderma_species,10)
print("Selected KOs: ")
print(pilo10)
piloCroceumData <- megaMeta[megaMeta$V1 %in% genes_piloderma_croceum,]
piloCroceumKOgenes <- unique(piloCroceumData[grepl(paste(pilo10,collapse="|"), piloCroceumData$V9),]$V1)

unkPiloData <- megaMeta[megaMeta$V1 %in% genes_unclassified_piloderma,]
unkPiloKOgenes <- unique(unkPiloData[grepl(paste(pilo10,collapse="|"), unkPiloData$V9),]$V1)
```

PILODERMA CROCEUM genes to KO profiles
```{r echo=FALSE, results='asis',message=FALSE, warning=FALSE,  fig.hold='hold', out.width="50%"}
p1 <- plotEigengene(controlData, intersect(piloCroceumKOgenes,colnames(controlData)), weeksC, treatmentC)

p2 <- plotEigengene(fertilisedData, intersect(piloCroceumKOgenes,colnames(fertilisedData)), weeksF, treatmentF, colors = "#BE9230")

cat("\n\n")
cat("\n\n")
par(mar = c(4, 4, 0.1, 0.1))
print(p1)
print(p2)
par(mfrow=c(1,1))
cat("\n\n")
cat("\n\n")
```

UNKNOWN PILODERMA genes to KO profiles
```{r echo=FALSE, fig.hold='hold', message=FALSE, warning=FALSE, , results='asis', out.width="50%"}
p1 <- plotEigengene(controlData, intersect(unkPiloKOgenes,colnames(controlData)), weeksC, treatmentC)

p2 <- plotEigengene(fertilisedData, intersect(unkPiloKOgenes,colnames(fertilisedData)), weeksF, treatmentF, colors = "#BE9230")

cat("\n\n")
cat("\n\n")
par(mar = c(4, 4, 0.1, 0.1))
print(p1)
print(p2)
par(mfrow=c(1,1))
cat("\n\n")
cat("\n\n")
```

# Session information
```{r}
sessionInfo()
```