---
  title: "Flakaliden spruce roots + fungi cluster enrichment analysis"
author: "Alonso Serrano"
date: "`r format(Sys.time(), '%d %B %Y')`"
output: 
  html_document: 
  number_sections: yes
toc: yes
---
  
  <style>
  * {
    box-sizing: border-box;
  }

/* Create two equal columns that floats next to each other */
  .column {
    float: left;
    width: 50%;
    padding: 10px;
    
  }

/* Clear floats after the columns */
  .row:after {
    content: "";
    display: table;
    clear: both;
  }
</style>
  
  ```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

# Introduction
This file contains the script and results of the data from spruce roots and fungi from the 2011 Flakaliden dataset. 

## Prerequisites
The variance stabilization transformed dataset, either as TSV or RData obtained from de differential expression analysis step.
Infomap clustering results from Seidr network
Pre enriched data using gofer

## Setup
```{r generalSetup, message=FALSE, warning=FALSE}
library(data.table)
library(dplyr)
library(here)

# preload packages for depencies
library(ggplot2)
library(RColorBrewer)
library(reshape2)
library(tidyr)
library(tibble)
library(jsonlite)
library(treemap)
library(KEGGREST)
library(reshape2)

source("~/Git/Rtoolbox/src/plotEigenGene.R")
source("~/Git/Rtoolbox/src/getEigengenes.R")
source("~/Git/Rtoolbox/src/plotEnrichedTreemap.R")
source("~/Git/UPSCb-common/src/R/gopher.R")
source("~/scripts/koPathwayDiseaesCleanup/koProcessing.R")
#source("corAuxFunctions.R")

spruceDir <- "/mnt/picea/projects/spruce/vhurry/14_SpruceRoots_Project"
fungiDir <- "/mnt/picea/projects/spruce/vhurry/fungi-flakaliden-2011"
cortinariusDir <- "/mnt/picea/projects/spruce/vhurry/fungi-flakaliden-2011/fungi-species/cortinarius_glaucopus"
cenococcumDir <- "/mnt/picea/projects/spruce/vhurry/fungi-flakaliden-2011/fungi-species/cenococcum_geophilum"

spruceDE <- file.path(spruceDir,"DE")
fungiDE <- file.path(fungiDir,"DE")
cortiDE <- file.path(cortinariusDir,"DE")
cenoDE <- file.path(cenococcumDir,"DE")

checkFile <- function(fileName) {
  if (!file.exists(fileName)) stop(paste("File", fileName, "doesn't exist"))
}
```

## Check files

### VSD data file
The files are tab separated, rows contain sample meta information (week and treatment),
columns contain gene or KEGG orthologs names. The values are expression data after variance stabilization transformation

### Cluster files
They must contain two columns gene and cluster and be tab separated<br />
  ex.<br />
  gene	cluster<br />
  MA_5320g0010	Cluster1<br />
  MA_32763g0010	Cluster1<br />
  MA_66378g0010	Cluster1<br />
  
  
```{r loadRdata}
load(file.path(spruceDE, "controlVsdData.RData"))
controlSpruceData <- controlData

load(file.path(fungiDE, "controlVsdData.RData"))
controlFungiData <- controlData

load(file.path(cortiDE, "controlVsdData.RData"))
controlCortiData <- controlData

load(file.path(cenoDE, "controlVsdData.RData"))
controlCenoData <- controlData
```

```{r}
# load spruce data files
controlSpruceClusterFile <- file.path(spruceDir,"spruce-networks/control/cluster","InfomapClusters.tsv")
checkFile(controlSpruceClusterFile)

# load fungi data files
controlFungiClusterFile <- file.path(fungiDir,"networks/control/cluster","InfomapClusters.tsv")
checkFile(controlFungiClusterFile)

controlCortiClusterFile <- file.path(cortinariusDir,"networks/control/results/aggregated","infomapClusters.tsv")
checkFile(controlCortiClusterFile)

controlCenoClusterFile <- file.path(cenococcumDir,"networks/control/results/aggregated","infomapClusters.tsv")
checkFile(controlCenoClusterFile)

```

## Load files
```{r loadFiles}
#name correction, spruce rownames, have an extra dash
rownames(controlSpruceData) <- gsub("-", "", rownames(controlSpruceData))
#rownames(fertilisedSpruceData) <- gsub("-", "", rownames(fertilisedSpruceData))

controlSpruceClusters <- read.table(controlSpruceClusterFile, stringsAsFactors=F, header=T, sep='\t', comment.char="")

controlFungiClusters <- read.table(controlFungiClusterFile, stringsAsFactors=F, header=T, sep='\t', comment.char="")

controlCortiClusters <- read.table(controlCortiClusterFile, stringsAsFactors=F, header=T, sep='\t', comment.char="")

controlCenoClusters <- read.table(controlCenoClusterFile, stringsAsFactors=F, header=T, sep='\t', comment.char="")
```

## Create metadata based on the information on sample names
```{r}
controlSpruceGroup <- substr(rownames(controlSpruceData),4,4)
controlSpruceTime <- as.integer(substr(rownames(controlSpruceData),2,3))

controlFungiGroup <- substr(rownames(controlFungiData),4,4)
controlFungiTime <- as.integer(substr(rownames(controlFungiData),2,3))

controlCortiGroup <- substr(rownames(controlCortiData),4,4)
controlCortiTime <- as.integer(substr(rownames(controlCortiData),2,3))

controlCenoGroup <- substr(rownames(controlCenoData),4,4)
controlCenoTime <- as.integer(substr(rownames(controlCenoData),2,3))

# fertilisedSpruceGroup <- substr(rownames(fertilisedSpruceData),4,4)
# fertilisedSpruceTime <- as.integer(substr(rownames(fertilisedSpruceData),2,3))
# 
# fertilisedFungiGroup <- substr(rownames(fertilisedFungiData),4,4)
# fertilisedFungiTime <- as.integer(substr(rownames(fertilisedFungiData),2,3))
```



Enrich & plot results


```{r, eval=T, include=F}
load(file=file.path(spruceDir,"spruce-networks/control/cluster/controlSplitEnr.RData"))

controlSpruceEnr <- controlSplitEnr
```



```{r, eval=F,include=F}
save(enrRes, file=here("env", paste0(specie, "_enrRes.RData")))
```





## Enrich & plot
```{r superLoop2, fig.hold='hold', message=FALSE, warning=FALSE, out.width="50%", results='asis'}
None <- lapply(1:20, function(x) {
  spruceClus <- mostDegraded[x,]$`Spruce-control`
  fungiClus <- mostDegraded[x,]$`Fungi-control`
  cat("\n")
  cat("<hr>")
  cat(paste0("<H3>Spruce cluster: ", spruceClus, ". Fungi cluster ",fungiClus, ". Degradation: ", round(mostDegraded[x,]$degradation, 4),"</H3>"))
  cat("\n\n")
  spruceMulti <- plotThis(controlSpruceData,  fertilisedSpruceData,
                          controlSpruceClusters,              controlSpruceGroup, 
                          controlSpruceTime, 
                          fertilisedSpruceGroup, 
                          fertilisedSpruceTime, 
                          spruceClus, 
                          
                          title1 = paste("Spruce control cluster", spruceClus, "with control data"),
                          title2 = paste("Spruce control cluster", spruceClus, "with fertilised data")
                          
  )
  
  fungiMulti <- plotThis(controlFungiData,  fertilisedFungiData,
                         controlFungiClusters, 
                         controlFungiGroup, 
                         controlFungiTime, 
                         fertilisedFungiGroup, 
                         fertilisedFungiTime, 
                         fungiClus, 
                         
                         title1 = paste("Fungi control cluster", fungiClus, "with control data"),
                         title2 = paste("Fungi control cluster", fungiClus, "with fertilised data")
  )
  
  cat(paste("<H4>Profiles plot</H4>"))
  cat("\n\n")
  cat(paste("<center>Spruce - fungi control profiles correlation",round(mostDegraded[x,]$control_score,4), "</center>"))
  cat("\n\n")
  par(mar = c(4, 4, 0.1, 0.1))
  print(spruceMulti$plot1)
  print(fungiMulti$plot1)
  cat(paste("<div class='row'>",
            "<div class='column'><center>Spruce correlation", round(spruceMulti$cor1, 4), "</center></div>",
            "<div class='column'><center>Fungi correlation", round(fungiMulti$cor1, 4), "</center></div></div>"
  ))
  
  print(spruceMulti$plot2)
  print(fungiMulti$plot2)
  par(mfrow=c(1,1))
  cat("\n\n")
  cat("\n\n")
  cat(paste("<center>Spruce - fungi fertilised profiles correlation",round(mostDegraded[x,]$fertilised_score,4),"</center>"))
  cat("\n\n")
  cat("\n\n")
  
  cat(paste("<H4>Spruce treemaps</H4>"))
  cat("\n\n")
  par(mar = c(4, 4, 0.1, 0.1))
  None <- sapply(c("go","pfam","mapman","kegg"), function(y){
    if(!is.null(enrRes[[x]]$spruce[[y]])) {
      if(y=="kegg"){
        enrRes[[x]]$spruce[[y]] <- koTranslate(enrRes[[x]]$spruce[[y]])
      }
      plotEnrichedTreemap(enrRes[[x]]$spruce, enrichment = y, namespace = "none", title=paste(y,"enrichment"))
    }
  })
  par(mfrow=c(1,1))
  cat("\n\n")
  cat("<hr>")
  
  cat(paste("<H4>Fungi treemaps</H4>"))
  cat("\n\n")
  par(mar = c(4, 4, 0.1, 0.1))
  
  None <- sapply(c("ko_pathway"), function(y){
    if(!is.null(enrRes[[x]]$fungi[[y]])) {
      if(y=="ko_pathway"){
        enrRes[[x]]$fungi[[y]] <- koPathwayDiseaseCleanup(koTranslate(enrRes[[x]]$fungi[[y]]))
      }
      plotEnrichedTreemap(enrRes[[x]]$fungi, enrichment = y, namespace = "none",title=paste(y,"enrichment"))
    }
    
  })
  par(mfrow=c(1,1))
  cat("\n\n")
  cat("<hr>")
  cat("<hr>")
})

```
```{r splitClusters}

controlFungiClusters

splitClusters <- lapply(unique(controlFungiClusters$cluster), function(cluster){
  print(cluster)
})

splitClusters <- lapply(clusterNames, function(cluster){
  theCluster <- InfomapClusters[InfomapClusters$cluster==cluster,]$gene
  cat(paste0("<H3>", cluster, "","</H3>"))
  cat("\n\n")
  res <- getEigenGenes(theData, theCluster)
  pos <-length(res$positive)
  neg <- length(res$negative)
  posRatio <- round((pos/(pos+neg)) * 100, digits = 2)
  
  ratio <- paste0(posRatio,"-",100-posRatio)
  cat(paste("Genes in main profile", pos, "\n"))
  cat(paste("Genes in reverse profile", neg, "\n"))
  cat(paste("Main-reverse ratio", ratio, "\n"))
  res
})
names(splitClusters) <- clusterNames
```

```{r, eval=F, include=F}
save.image(file.path(here("env"),"env.RData"))
```

# Session information
```{r}
sessionInfo()
```
