---
title: "Fungi Flakaliden project - Comparative year analysis"
author: "Alonso Serrano"
date: "`r Sys.Date()`"
output:
 html_document:
   toc: true
   number_sections: true
---

# Introduction
This file contains the script and results for the comparatived analysis done to the sfungi Flakaliden 2011 & 2012 data. 

*200401 Creation  
*200407 Added new yearly DE section  
*200409 Added Comparition F vs C within same timepoint in different years  

## Prerequisites
All RNAseq Fastq files where previously processed by a nextflow pipeline for RNAseq analysis.
The files where QA with fastQC, then trimmed by Trimmomatic and sortmeRNA and finally aligned by
Salmon.
The output of the pipeline are the Salmon files, one per original fastq file.

Metadata files containing all the information related to each sample, file name, time, condition and replica number.

Vst and DESeqMAtrix objects from 2011 and 2012 fungi data.

# Setup
Load libraries
```{r, message=FALSE}
library(DESeq2)
library(tximport)
library(BiocParallel)
library(plotly)
library(dplyr)
library(here)
library(limma)
library(VennDiagram)
library(gridExtra)
register(MulticoreParam(4))
source(here("Rtoolbox/src/utilsDE.r"))
source(here("Rtoolbox/src/plot3Dvector.R"))
source(here("Rtoolbox/src/plotVectorPCA.R"))
source(here("Rtoolbox/src/plotUpAndDown.R"))
```

Set the project folder
```{r}
project2011Folder <- "/mnt/picea/projects/spruce/vhurry/fungi-flakaliden-2011"
de2011Folder <- file.path(project2011Folder, "DE")
data2011Folder <- file.path(project2011Folder, "data")

project2012Folder <- "/mnt/picea/projects/spruce/vhurry/fungi-flakaliden-2012"
de2012Folder <- file.path(project2012Folder, "DE")
data2012Folder <- file.path(project2012Folder, "data")
```

```{r message=TRUE, warning=TRUE, include=FALSE}
load(here("env", "allDEFungiMergedResults.RData"))

```

Load yearly datasets
```{r loadVSTpreprocessed, echo=FALSE}
meta.2011 <- read.table(file.path(data2011Folder, "meta.tsv"), header = T,
                        sep='\t', comment.char = "", stringsAsFactors = F)

kos.2011 <- read.table(file.path(data2011Folder, "kos.raw.tsv"), header = T,
                       sep='\t', comment.char = "",  quote="",stringsAsFactors = F)

meta.2012 <- read.table(file.path(data2012Folder, "meta.tsv"), header = T,
                        sep=';', comment.char = "", stringsAsFactors = F)

kos.2012 <- read.table(file.path(data2012Folder, "kos.raw.tsv"), header = T,
                       sep='\t', comment.char = "", quote="",stringsAsFactors = F)
```

## Dataset merge

Adapt 2011 meta
```{r}
meta.2011.red <- data.frame(id=meta.2011$Sample.ID,
                            treatment=as.character(meta.2011$Condition),
                            date=as.character(meta.2011$Sampling.date..week..),
                            year=rep("2011", length(meta.2011$Sample.ID)),
                            stringsAsFactors = F)
meta.2011.red$treatment[meta.2011.red$treatment == 'control'] <- 'Control'
meta.2011.red$treatment[meta.2011.red$treatment == 'treatment'] <- 'Fertilised'
meta.2011.red$date <- gsub("W","",meta.2011.red$date)
meta.2011.red$sample <- paste("2011", meta.2011.red$treatment, 
                              meta.2011.red$date, sep = "_")
```

Adapt 2012 meta
```{r}
meta.2012.red <- data.frame(id=meta.2012$SciLifeID,
                            treatment=as.character(meta.2012$treatment),
                            date=as.character(meta.2012$date),
                             year=rep("2012", length(meta.2012$SciLifeID)),
                            stringsAsFactors = F)

meta.2012.red$date[meta.2012.red$date == 'Early_June'] <- '23'
meta.2012.red$date[meta.2012.red$date == 'Late_June'] <- '25'
meta.2012.red$date[meta.2012.red$date == 'August'] <- '32'
meta.2012.red$date[meta.2012.red$date == 'October'] <- '41'

meta.2012.red$treatment <- ifelse(meta.2012.red$treatment=="25_year", "Fertilised", meta.2012.red$treatment)
meta.2012.red <- meta.2012.red[meta.2012.red$treatment != "5_year",]
meta.2012.red$treatment <- factor(meta.2012.red$treatment, levels = c("Control", "Fertilised") )


meta.2012.red$sample <- paste("2012", meta.2012.red$treatment, 
                              meta.2012.red$date, sep = "_")
```

Merge metadata
```{r mergeMeta}
meta <- rbind(meta.2011.red, meta.2012.red)
```

```{r}
kos.2011 <- kos.2011[kos.2011$KO_name !='Unclassified',]
kos.2012 <- kos.2012[kos.2012$KO_name !='Unclassified',]


rownames(kos.2011) <- substr(kos.2011[,1],1,6)
kos.2011 <- kos.2011[,-1]
colnames(kos.2011)<-gsub("X","",colnames(kos.2011))

rownames(kos.2012) <- substr(kos.2012[,1],1,6)
kos.2012 <- kos.2012[,-1]
```

```{r}
matrix.2011 = kos.2011[,which(colnames(kos.2011) %in% meta.2011.red$id)]
matrix.2011 <- matrix.2011[, match(meta.2011.red$id, colnames(matrix.2011))]
all(colnames(matrix.2011) == meta.2011.red$id)
colnames(matrix.2011) <- meta.2011.red$sample


matrix.2012 = kos.2012[,which(colnames(kos.2012) %in% meta.2012.red$id)]
matrix.2012 <- matrix.2012[, match(meta.2012.red$id, colnames(matrix.2012))]
all(colnames(matrix.2012) == meta.2012.red$id)
colnames(matrix.2012) <- meta.2012.red$sample

```

Both datasets contain different number of KO, a Venn Diagramm will show how many genes are exclusive
```{r}
grid.newpage()
draw.pairwise.venn(length(rownames(matrix.2011)),
                   length(rownames(matrix.2012)), 
                   length(intersect(rownames(matrix.2011), rownames(matrix.2012))),
                   category = c("2011", "2012"), 
                   lty = rep("blank", 2), 
                   fill = c("orange", "red"), 
                   alpha = rep(0.5, 2), 
                   cat.pos = c(0, 0), 
                   cat.dist = rep(0.025, 2), 
                   scaled = FALSE)
```

2012 data contains a big amount of exclusive KO, but it seems that many of them are very low abundance KOs. A zeroMAD cleanup will delete all "noisy" KOs.
```{r}
matrix.2011 <- matrix.2011[rowMads(as.matrix(matrix.2011)) > 0,]
matrix.2012 <- matrix.2012[rowMads(as.matrix(matrix.2012)) > 0,]
```

We replot the Venn diagram to check if it is coherent
```{r}
grid.newpage()
draw.pairwise.venn(length(rownames(matrix.2011)),
                   length(rownames(matrix.2012)), 
                   length(intersect(rownames(matrix.2011), rownames(matrix.2012))),
                   category = c("2011", "2012"), 
                   lty = rep("blank", 2), 
                   fill = c("orange", "red"), 
                   alpha = rep(0.5, 2), 
                   cat.pos = c(0, 0), 
                   cat.dist = rep(0.025, 2), 
                   scaled = FALSE)
```

The Venn Diagram shows low exclusive KOs for each dataset, these can be due to the difference in the number of timepoints and/or number of replicates per timepoint.
We obtain the list of KOs in common
```{r}
kos <- intersect(rownames(matrix.2011), rownames(matrix.2012))
kos <- sort(kos)
```

Create the subsetted matrix from the original datasets.
```{r}
matrix.2011.red <- matrix.2011[rownames(matrix.2011) %in% kos,]
matrix.2012.red <- matrix.2012[rownames(matrix.2012) %in% kos,]
```

Check the dimensions
```{r checkDim}
print(dim(matrix.2011.red))
print(dim(matrix.2012.red))
print(dim(meta.2011.red))
print(dim(meta.2012.red))
```

Merge the datasets
```{r}
matrix <- cbind(matrix.2011.red, matrix.2012.red)
```

Import into a DESeqDataSet
```{r}
ddsMatrix <- DESeqDataSetFromMatrix(matrix, colData = meta, design=~sample)
colData(ddsMatrix)$year <- factor(colData(ddsMatrix)$year)
```

# QA
## PCA
We use blind vst to plot a PCA and check for patterns and outliers

```{r}
vsd.QA <- vst(ddsMatrix, blind=TRUE)
geneNumber <- length(rownames(assay(vsd.QA)))
```

```{r}
plotPCA(vsd.QA, intgroup = c("treatment"), ntop=geneNumber)
```

```{r}
plotPCA(vsd.QA, intgroup = c("date"), ntop=geneNumber)
```

```{r}
plotPCA(vsd.QA, intgroup = c("year"), ntop=geneNumber)
```

## PCA with year effect removed
We use blind vst to plot a PCA and check for patterns and outliers
```{r}
vsd.fix <- vsd.QA
assay(vsd.fix) <- limma::removeBatchEffect(assay(vsd.fix), vsd.fix$year)
```

```{r}
geneNumber <- length(rownames(assay(vsd.fix)))
```

```{r}
plotPCA(vsd.fix, intgroup = c("treatment"), ntop=geneNumber)
```
```{r}
plotPCA(vsd.fix, intgroup = c("date"), ntop=geneNumber)
```

```{r}
plotPCA(vsd.fix, intgroup = c("year"), ntop=geneNumber)
```

## 3D PCA
We represent the first 3 compoments as a 3D interactive object
```{r}

vsd.QA.t <- t(assay(vsd.fix))
pca <- prcomp(vsd.QA.t)
percents <- round(summary(pca)$importance[2,]*100)

colors <- c("#f8766d", "#7cae00","#00bfc4")

df3D <- data.frame(treatment = meta$treatment)
df3D$time <- meta$date
df3D$colors <- colors[as.numeric(df3D$treatment)]
df3D$PC1 <- pca$x[,1]
df3D$PC2 <- pca$x[,2]
df3D$PC3 <- pca$x[,3] 

plot_ly(df3D, x = ~PC1, y = ~PC3, z = ~PC2, 
        color = ~treatment, colors = colors,
        text = ~paste("Date:", time, '<br>Treatment:', treatment) ) %>%   
  add_markers() %>%
  layout(scene = list(xaxis = list(title = paste("PC1 (",percents[1],"%)",sep="")),
                      yaxis = list(title = paste("PC3 (",percents[3],"%)",sep="")),
                      zaxis = list(title = paste("PC2 (",percents[2],"%)",sep=""))))

```

# DE comparition

Load precalculated DE results from both years
```{r}
load(file.path(de2011Folder, "allDEresults.RData"))
control.2011.res <- control.res
fertilised.2011.res <- fertilised.res
combined.2011.res <- combined.res

load(file.path(de2012Folder, "allDEresults.RData"))
control.2012.res <- control.res
fertilised.2012.res <- fertilised.res
combined.2012.res <- combined.res
```

Using the filterDE function we apply the padj of 0.01 and an absolute log2fc of 0.5
```{r filterStage, results="hide", message=FALSE}
control.2011.res.filter <- lapply(control.2011.res, filterDE)
fertilised.2011.res.filter <- lapply(fertilised.2011.res, filterDE)
combined.2011.res.filter <- lapply(combined.2011.res, filterDE)

control.2012.res.filter <- lapply(control.2012.res, filterDE)
fertilised.2012.res.filter <- lapply(fertilised.2012.res, filterDE)
combined.2012.res.filter <- lapply(combined.2012.res, filterDE)
```

## Plot DE genes
We can plot the results as positive and negative bars to reflect the amount of genes up or down
regulated in each comparition 

### Control date+1 vs date plot
```{r plotControl, echo=FALSE, results='asis'}
plotUpAndDown(control.2011.res.filter, xlabel = "2011 comparitions")
plotUpAndDown(control.2012.res.filter, xlabel = "2012 comparitions")
```

### Fertilised date+1 vs date plot
```{r plotFert}
plotUpAndDown(fertilised.2011.res.filter, xlabel = "2011 comparitions")
plotUpAndDown(fertilised.2012.res.filter, xlabel = "2012 comparitions")
```

### Fertilised vs Control per date plot
```{r plotCombined}
plotUpAndDown(combined.2011.res.filter, xlabel = "2011 comparitions")
plotUpAndDown(combined.2012.res.filter, xlabel = "2012 comparitions")
```

# Merged Differential Expression common timepoints
The dataset is ready for DESeq2. We check the design of the DE object 

```{r}
design(ddsMatrix)
```

We are interested on future comparitions like DE genes in each timepoints or between timepoints
that is why we created the interaction group

## Run DESeq
```{r runDE, eval=FALSE}
dds <- DESeq(ddsMatrix)
```

We check the result names
```{r}
resultsNames(dds)
```

## DE between years in common conditions and common timepoints

The common timepoint are weeks: 23, 25, 32 and 41
In order to make all the comparitions automated, we prepare vectors with the elements that will be used
in the result extraction, we generate 4 vectors, one per condition + year and we narrow it to the timepoints
of interest
```{r}
commonTimepoints <- c("23", "25", "32", "41")

control2011Vector <- as.character(unique(meta.2011.red$sample[grep("Control", meta.2011.red$sample)]))
control2011Vector <- sapply(commonTimepoints, function(x){control2011Vector[grep(x, control2011Vector)]})

control2012Vector <- as.character(unique(meta.2012.red$sample[grep("Control", meta.2012.red$sample)]))
control2012Vector <- sapply(commonTimepoints, function(x){control2012Vector[grep(x, control2012Vector)]})

fertilised2011Vector <- as.character(unique(meta.2011.red$sample[grep("Fertilised", meta.2011.red$sample)]))
fertilised2011Vector <- sapply(commonTimepoints, function(x){fertilised2011Vector[grep(x, fertilised2011Vector)]})

fertilised2012Vector <- as.character(unique(meta.2012.red$sample[grep("Fertilised", meta.2012.red$sample)]))
fertilised2012Vector <- sapply(commonTimepoints, function(x){fertilised2012Vector[grep(x, fertilised2012Vector)]})
```

## Result extraction

### Control date vs date

For this comparition we take consider only the control condition and we compare each pair of weeks
to determine which genes are up or down regaluated from one week to the previous week

```{r message=TRUE, warning=TRUE, include=FALSE}
load(here("env", "allDEFungiMergedResults.RData"))

```

```{r eval=FALSE}
control.res <- mapply(getRes, control2012Vector,
                      control2011Vector, 
                      MoreArgs = list(localDDS=dds, group="sample"))

fertilised.res <- mapply(getRes, fertilised2012Vector,
                      fertilised2011Vector, 
                      MoreArgs = list(localDDS=dds, group="sample"))


```

We add the corresponding names to the results
```{r eval=FALSE}
names(control.res) <- c("2012_Control_23 vs 2011_Control_23", 
                        "2012_Control_25 vs 2011_Control_25", 
                        "2012_Control_32 vs 2011_Control_32",
                        "2012_Control_41 vs 2011_Control_41")
names(fertilised.res) <- c("2012_Fertilised_23 vs 2011_Fertilised_23", 
                        "2012_Fertilised_25 vs 2011_Fertilised_25", 
                        "2012_Fertilised_32 vs 2011_Fertilised_32",
                        "2012_Fertilised_41 vs 2011_Fertilised_41")

```

The results contain all the genes, we need to filter them to get only those genes
that are stadistical significant and that have a strong condtition effect
For that we discard all the genes that has a padj higher than 0.01 and
a log2FoldChange lower than 0.5 (absolute value)

```{r filtering, results="hide", message=FALSE}
control.res.filter <- lapply(control.res, filterDE)
fertilised.res.filter <- lapply(fertilised.res, filterDE)
```

## Plot DE genes

We can plot the results as positive and negative bars to reflect the amount of genes up or down
regulated in each comparition 

### Control 2012 vs 2011 plot

```{r}
plotUpAndDown(control.res.filter)
```

### Fertilised 2012 vs 2011 plot

```{r}
plotUpAndDown(fertilised.res.filter)
```

## DE between Fertilised and Control individual years

We subset the combined results (Fertilised vs control) so both datasets have the same sampling point

```{r plotCombinedRed}
combined.2011.res.filter.red <- combined.2011.res.filter[c("23Fvs23C", "25Fvs25C", "32Fvs32C", "41Fvs41C")]
combined.2012.res.filter.red <- combined.2012.res.filter
names(combined.2012.res.filter.red) <- c("23Fvs23C", "25Fvs25C", "32Fvs32C", "41Fvs41C")
```

Plot to see the DE filtered genes

```{r}
plotUpAndDown(combined.2011.res.filter.red, xlabel = "2011 comparitions")
plotUpAndDown(combined.2012.res.filter.red, xlabel = "2012 comparitions")
```

Generate Venn diagrams to see the overlap between up and down regulated genes within each timepoint.

```{r}
timepoints <- c("23Fvs23C", "25Fvs25C", "32Fvs32C", "41Fvs41C")

sapply(timepoints, function(timepoint) {
  up2011 <- rownames(combined.2011.res.filter.red[[timepoint]][combined.2011.res.filter.red[[timepoint]]$log2FoldChange>0,])
  down2011 <- rownames(combined.2011.res.filter.red[[timepoint]][combined.2011.res.filter.red[[timepoint]]$log2FoldChange<0,])
  
  up2012 <- rownames(combined.2012.res.filter.red[[timepoint]][combined.2012.res.filter.red[[timepoint]]$log2FoldChange>0,])
  down2012 <- rownames(combined.2012.res.filter.red[[timepoint]][combined.2012.res.filter.red[[timepoint]]$log2FoldChange<0,])
  
  g <- draw.quad.venn(area1 = length(up2011), 
                        area2 = length(down2011), 
                        area3 = length(up2012),
                        area4 = length(down2012),
                        n12 = length(intersect(up2011, down2011)), 
                        n13 = length(intersect(up2011, up2012)),
                        n14 = length(intersect(up2011, down2012)),
                        n23 = length(intersect(down2011, up2012)), 
                        n24=length(intersect(down2011, down2012)),
                        n34=length(intersect(up2012, down2012)),
                        n123=length(intersect(up2011, intersect(down2011, up2012))), 
                        n124=length(intersect(up2011, intersect(down2011, down2012))), 
                        n134=length(intersect(up2011, intersect(up2012, down2012))), 
                        n234=length(intersect(down2011, intersect(up2012, down2012))),
                        n1234=length(intersect(intersect(up2011, down2011),intersect(up2012, down2012))),
                          category = c("Up 2011", "Down 2011", "Up 2012", "Down 2012"),
                         fill = c("orange", "red", "green", "blue"),
                          euler.d = TRUE, scaled = FALSE,
               cat.col = c("orange", "red", "green", "blue"),
               ind=F,
                          filename = NULL
                          )


  grid.arrange(gTree(children=g), top="Up & down regulated DE genes", bottom=paste("Comparition", timepoint))

    ### Averaging the log2foldchange of significant DE genes

  lfcList.red <- list(up2011 = combined.2011.res.filter.red[[timepoint]][combined.2011.res.filter.red[[timepoint]]$log2FoldChange>0,]$log2FoldChange,
                      up2012 = combined.2012.res.filter.red[[timepoint]][combined.2012.res.filter.red[[timepoint]]$log2FoldChange>0,]$log2FoldChange,
                      down2011 = combined.2011.res.filter.red[[timepoint]][combined.2011.res.filter.red[[timepoint]]$log2FoldChange<0,]$log2FoldChange,
                      down2012 = combined.2012.res.filter.red[[timepoint]][combined.2012.res.filter.red[[timepoint]]$log2FoldChange<0,]$log2FoldChange )
  
  boxplot(lfcList.red, ylim=c(-5,5), col=c("#00FF00", "#FF0000"), 
          ylab="log2FoldChange",
          main = paste("lfc distribution for: ", timepoint))
})

```

# Merged Differential Expression year effect

We need to change the design of the DE object to focus only on year

```{r}
design(ddsMatrix) <- ~year
```

We are interested on future comparitions like DE genes in each timepoints or between timepoints
that is why we created the interaction group

## Run DESeq
```{r runDEYear, eval=FALSE}
ddsYear <- DESeq(ddsMatrix)
```

We check the result names
```{r}
resultsNames(ddsYear)
```

Obtain the results
```{r evalYear}
year.res <- list(results(ddsYear ,
        name= "year_2012_vs_2011",
        filter = rowMedians(counts(ddsYear )),
        parallel = FALSE))
names(year.res) <- "2012vs2011"
```

Filter the results
```{r filterYear, results="hide", message=FALSE}
year.res.filter <- lapply(year.res, filterDE)
```

### 2012 vs 2011 plot
```{r}
plotUpAndDown(year.res.filter)
```

```{r saveresults,eval=FALSE, echo=FALSE}
save(dds, ddsYear, control.res, fertilised.res, year.res, 
     file=here("env", "allDEFungiMergedResults.RData"))
```

## DE between Fertilised and Control comparitions in each dataset
```{r}
load(file.path(de2011Folder, "ddsTreatment.RData"))
dds2011FungiTreatment <- ddsFungiTreatment
res2011FungiTreatment <- resFungiTreatment
res2011FungiTreatment.filter <- lapply(res2011FungiTreatment, filterDE)

load(file.path(de2012Folder, "ddsTreatment.RData"))
dds2012FungiTreatment <- ddsFungiTreatment
res2012FungiTreatment <- resFungiTreatment
res2012FungiTreatment.filter <- lapply(res2012FungiTreatment, filterDE)

```

2011 Treatment vs Control filtered plot
```{r}
plotUpAndDown(res2011FungiTreatment.filter)
```

2012 Treatment vs Control filtered plot
```{r}
plotUpAndDown(res2012FungiTreatment.filter)
```

```{r}
Fungi2011Up <- rownames(res2011FungiTreatment.filter$FertilisedvsControl[res2011FungiTreatment.filter$FertilisedvsControl$log2FoldChange>0,])
Fungi2011Down <- rownames(res2011FungiTreatment.filter$FertilisedvsControl[res2011FungiTreatment.filter$FertilisedvsControl$log2FoldChange<0,])

Fungi2012Up <- rownames(res2012FungiTreatment.filter$FertilisedvsControl[res2012FungiTreatment.filter$FertilisedvsControl$log2FoldChange>0,])
Fungi2012Down <- rownames(res2012FungiTreatment.filter$FertilisedvsControl[res2012FungiTreatment.filter$FertilisedvsControl$log2FoldChange<0,])

FungiUp <- intersect(Fungi2011Up, Fungi2012Up)
FungiDown <- intersect(Fungi2011Down, Fungi2012Down)
```

Up-regulated genes in common in both years
```{r}
grid.newpage()
draw.pairwise.venn(length(Fungi2011Up),
                   length(Fungi2012Up), 
                   length(FungiUp),
                          category = c("2011", "2012"), lty = rep("blank", 
    2), fill = c("light blue", "pink"), alpha = rep(0.5, 2), cat.pos = c(0, 
    0), cat.dist = rep(0.025, 2), scaled = FALSE)
```

Down-regulated genes in common in both years
```{r}
grid.newpage()
draw.pairwise.venn(length(Fungi2011Down),
                   length(Fungi2012Down), 
                   length(FungiDown),
                          category = c("2011", "2012"), lty = rep("blank", 
    2), fill = c("light blue", "pink"), alpha = rep(0.5, 2), cat.pos = c(0, 
    0), cat.dist = rep(0.025, 2), scaled = FALSE)
```
### Averaging the log2foldchange of significant DE genes
```{r}
lfcList <- list(Fungi11Up = res2011FungiTreatment.filter$FertilisedvsControl[res2011FungiTreatment.filter$FertilisedvsControl$log2FoldChange>0,]$log2FoldChange,
                Fungi12Up = res2012FungiTreatment.filter$FertilisedvsControl[res2012FungiTreatment.filter$FertilisedvsControl$log2FoldChange>0,]$log2FoldChange,
                Fungi11Down = res2011FungiTreatment.filter$FertilisedvsControl[res2011FungiTreatment.filter$FertilisedvsControl$log2FoldChange<0,]$log2FoldChange,
                Fungi12Down = res2012FungiTreatment.filter$FertilisedvsControl[res2012FungiTreatment.filter$FertilisedvsControl$log2FoldChange<0,]$log2FoldChange )

boxplot(lfcList, ylim=c(-5,5), col=c("#00FF00", "#FF0000"), ylab="log2FoldChange" )
```


```{r}

matrix.2011.mean <- sapply(unique(colnames(matrix.2011.red)), function(x)  rowMeans(matrix.2011.red[colnames(matrix.2011.red) %in% x]))
matrix.2011.mean <- floor(matrix.2011.mean)
matrix.2011.mean <- matrix.2011.mean[,c("2011_Fertilised_23", "2011_Fertilised_25", "2011_Fertilised_32", "2011_Fertilised_41", "2011_Control_23",
                      "2011_Control_25", "2011_Control_32", "2011_Control_41")]

matrix.2011.mean <-data.frame(matrix.2011.mean, stringsAsFactors = F)
colnames(matrix.2011.mean) <- gsub("X2011_","",colnames(matrix.2011.mean))

matrix.2012.mean <- sapply(unique(colnames(matrix.2012.red)), function(x)  rowMeans(matrix.2012.red[colnames(matrix.2012.red) %in% x]))
matrix.2012.mean <- floor(matrix.2012.mean)
matrix.2012.mean <-data.frame(matrix.2012.mean, stringsAsFactors = F)
colnames(matrix.2012.mean) <- gsub("X2012_","",colnames(matrix.2012.mean))

comparitions <- sort(colnames(matrix.2011.mean))

panel.cor <- function(x, y){
    usr <- par("usr"); on.exit(par(usr))
    par(usr = c(0, 1, 0, 1))
    r <- round(cor(x, y), digits=2)
    txt <- paste0("R = ", r)
    cex.cor <- 0.8/strwidth(txt)
    text(0.5, 0.5, txt, cex = cex.cor * r)
}
```

```{r}
pairs(data.frame(matrix.2011.mean$Control_25, matrix.2012.mean$Control_25),  lower.panel = panel.cor)
pairs(data.frame(matrix.2011.mean$Fertilised_25, matrix.2012.mean$Fertilised_25),lower.panel = panel.cor)
```
```{r}
pairs(data.frame(matrix.2011.mean$Control_32, matrix.2012.mean$Control_32),  lower.panel = panel.cor)
pairs(data.frame(matrix.2011.mean$Fertilised_32, matrix.2012.mean$Fertilised_32),  lower.panel = panel.cor)
```
```{r}
pairs(data.frame(matrix.2011.mean$Control_41, matrix.2012.mean$Control_41),  lower.panel = panel.cor)
pairs(data.frame(matrix.2011.mean$Fertilised_41, matrix.2012.mean$Fertilised_41),  lower.panel = panel.cor)
```

# Top 5 species comparative PCAs
```{r}
load(file.path(data2011Folder, "megaData.RData"))
megaData.2011 <- megaData
load(file.path(data2012Folder, "megaData.RData"))
megaData.2012 <- megaData


```
Selected species: Piloderma croceum, Cortinarius glaucopus, Cenococcum geophilum, Hyaloscypha variabilis and Hyaloscypha bicolor
```{r, results='asis'}
getComparativePCA <- function(species, megaData.2011, megaData.2012) {
  subData11 <- megaData.2011[megaData.2011$species == species,][,2:112]
  subData12 <- megaData.2012[megaData.2012$species == species,][,2:72]
  
  minimeta.2011 <- data.frame(treatment= substr(colnames(subData11), 4,4),
                              date= substr(colnames(subData11), 2,3))
  minimeta.2012 <- data.frame(treatment= substr(colnames(subData12), 4,4),
                              date= substr(colnames(subData12), 2,3))
  ddsMatrix.2011 <- DESeqDataSetFromMatrix(subData11, colData = minimeta.2011, design=~1)
  ddsMatrix.2012 <- DESeqDataSetFromMatrix(subData12, colData = minimeta.2012, design=~1)
  
  # ddsMatrix.2011 <- ddsMatrix.2011[(rowSums(counts(ddsMatrix.2011)) > 1),]
  # ddsMatrix.2012 <- ddsMatrix.2012[(rowSums(counts(ddsMatrix.2012)) > 1),]
  
  cts <- counts(ddsMatrix.2011)
  geoMeans <- apply(cts, 1, function(row) if (all(row == 0)) 0 else exp(sum(log(row[row != 0]))/length(row)) )
  ddsMatrix.2011 <- estimateSizeFactors(ddsMatrix.2011, geoMeans=geoMeans)
  
  cts <- counts(ddsMatrix.2012)
  geoMeans <- apply(cts, 1, function(row) if (all(row == 0)) 0 else exp(sum(log(row[row != 0]))/length(row)) )
  ddsMatrix.2012 <- estimateSizeFactors(ddsMatrix.2012, geoMeans=geoMeans)
  
  vsd.QA.2011 <- varianceStabilizingTransformation(ddsMatrix.2011, blind=TRUE)
  geneNumber.2011 <- length(rownames(assay(vsd.QA.2011)))
  
  vsd.QA.2012 <- varianceStabilizingTransformation(ddsMatrix.2012, blind=TRUE)
  geneNumber.2012 <- length(rownames(assay(vsd.QA.2012)))
  cat(paste0("<H3>2011 PCA - Treatment</H3>"))
  cat("\n\n")
  print(plotPCA(vsd.QA.2011, intgroup = c("treatment"), ntop=geneNumber.2011))
  cat(paste0("<H3>2012 PCA - Treatment</H3>"))
  cat("\n\n")
  print(plotPCA(vsd.QA.2012, intgroup = c("treatment"), ntop=geneNumber.2012))
  cat("\n\n")
  cat(paste0("<H3>2011 PCA - Date</H3>"))
  cat("\n\n")
  print(plotPCA(vsd.QA.2011, intgroup = c("date"), ntop=geneNumber.2011))
  cat(paste0("<H3>2012 PCA - Date</H3>"))
  cat("\n\n")
  print(plotPCA(vsd.QA.2012, intgroup = c("date"), ntop=geneNumber.2012))
  cat("\n\n")
}
```

## Piloderma croceum
```{r message=F, warning=F, results='asis'}
getComparativePCA("Piloderma croceum", megaData.2011, megaData.2012)
```

## Cortinarius glaucopus
```{r message=F, warning=F, results='asis'}
getComparativePCA("Cortinarius glaucopus", megaData.2011, megaData.2012)
```

## Cenococcum geophilum
```{r message=F, warning=F, results='asis'}
getComparativePCA("Cenococcum geophilum", megaData.2011, megaData.2012)
```

## Hyaloscypha variabilis
```{r message=F, warning=F, results='asis'}
getComparativePCA("Hyaloscypha variabilis", megaData.2011, megaData.2012)
```

## Hyaloscypha bicolor
```{r message=F, warning=F, results='asis'}
getComparativePCA("Hyaloscypha bicolor", megaData.2011, megaData.2012)
```


# Session information
```{r}
sessionInfo()
```