---
title: "Spruce roots Flakaliden project - Comparative year analysis"
author: "Alonso Serrano"
date: "`r Sys.Date()`"
output:
 html_document:
   toc: true
   number_sections: true
---

# Introduction
This file contains the script and results for the comparatived analysis done to the spruce roots Flakaliden 2011 & 2012 data. 

*200331 Creation  
*200407 Added new yearly DE section  
*200409 Added Comparition F vs C within same timepoint in different years  

## Prerequisites
All RNAseq Fastq files where previously processed by a nextflow pipeline for RNAseq analysis.
The files where QA with fastQC, then trimmed by Trimmomatic and sortmeRNA and finally aligned by
Salmon.
The output of the pipeline are the Salmon files, one per original fastq file.

Metadata files containing all the information related to each sample, file name, time, condition and replica number.

Vst and DESeqMatrix objects from 2011 and 2012 spruce data.

# Setup
Load libraries
```{r, message=FALSE}
library(DESeq2)
library(tximport)
library(BiocParallel)
library(plotly)
library(dplyr)
library(here)
library(limma)
library(grid)
library(VennDiagram)
library(gridExtra)
register(MulticoreParam(4))
source(here("Rtoolbox/src/utilsDE.r"))
source(here("Rtoolbox/src/plot3Dvector.R"))
source(here("Rtoolbox/src/plotVectorPCA.R"))
source(here("Rtoolbox/src/plotUpAndDown.R"))
```

Set the project folder
```{r}
project2011Folder <- "/mnt/picea/projects/spruce/vhurry/14_SpruceRoots_Project"
de2011Folder <- file.path(project2011Folder, "DE")

project2012Folder <- "/mnt/picea/projects/spruce/vhurry/spruce-flakaliden-root-2012"
de2012Folder <- file.path(project2012Folder, "DE")
```

```{r message=TRUE, warning=TRUE, include=FALSE}
load(here("env", "allDESpruceMergedResults.RData"))
```

Load yearly datasets
```{r loadVSTpreprocessed, echo=FALSE}
load(file.path(de2011Folder, "vsd_QA.RData"))
vsd.QA.2011 <- vsd.QA
meta.2011 <- meta
load(file.path(de2011Folder, "ddsMatrix_raw.RData"))
ddsMatrix.2011 <- ddsMatrix

load(file.path(de2012Folder, "vsd_QA.RData"))
vsd.QA.2012 <- vsd.QA
meta.2012 <- meta
load(file.path(de2012Folder, "ddsMatrix_raw.RData"))
ddsMatrix.2012 <- ddsMatrix
```

## Dataset merge

Adapt 2011 meta
```{r}
meta.2011.red <- data.frame(id=meta.2011$Sample,
                            treatment=as.character(meta.2011$Treatment),
                            date=as.character(meta.2011$Week),
                            year=rep("2011", length(meta.2011$Sample)),
                            stringsAsFactors = F)
meta.2011.red$treatment[meta.2011.red$treatment == 'C'] <- 'Control'
meta.2011.red$treatment[meta.2011.red$treatment == 'F'] <- 'Fertilised'
meta.2011.red$date <- gsub("W","",meta.2011.red$date)
meta.2011.red$sample <- paste("2011", meta.2011.red$treatment, 
                              meta.2011.red$date, sep = "_")
```

Adapt 2012 meta
```{r}
meta.2012.red <- data.frame(id=meta.2012$SciLifeID,
                            treatment=as.character(meta.2012$treatment),
                            date=as.character(meta.2012$date),
                             year=rep("2012", length(meta.2012$SciLifeID)),
                            stringsAsFactors = F)

meta.2012.red$date[meta.2012.red$date == 'Early_June'] <- '23'
meta.2012.red$date[meta.2012.red$date == 'Late_June'] <- '25'
meta.2012.red$date[meta.2012.red$date == 'August'] <- '32'
meta.2012.red$date[meta.2012.red$date == 'October'] <- '41'

meta.2012.red$sample <- paste("2012", meta.2012.red$treatment, 
                              meta.2012.red$date, sep = "_")
```

Merge metadata
```{r mergeMeta}
meta <- rbind(meta.2011.red, meta.2012.red)
```

Expression data
```{r}
matrix.2011 <- assay(ddsMatrix.2011)
all(colnames(matrix.2011) == meta.2011.red$id)
colnames(matrix.2011) <- meta.2011.red$sample

```

```{r}
matrix.2012 <- assay(ddsMatrix.2012)
all(colnames(matrix.2012) == meta.2012.red$id)
colnames(matrix.2012) <- meta.2012.red$sample
```

Both datasets contain different number of genes, a Venn Diagramm will show how many genes are exclusive
```{r}
matrix.2011b <- matrix.2011[rowMads(as.matrix(matrix.2011)) > 0,]
matrix.2012b <- matrix.2012[rowMads(as.matrix(matrix.2012)) > 0,]
grid.newpage()
draw.pairwise.venn(length(rownames(matrix.2011b)),
                   length(rownames(matrix.2012b)), 
                   length(intersect(rownames(matrix.2011b), rownames(matrix.2012b))),
                          category = c("2011", "2012"), lty = rep("blank", 
    2), fill = c("orange", "red"), alpha = rep(0.5, 2), cat.pos = c(0, 
    0), cat.dist = rep(0.025, 2), scaled = FALSE)
```

Genes in common
```{r}
genes <- intersect(rownames(ddsMatrix.2011), rownames(ddsMatrix.2012))

genes <- sort(genes)
```

```{r}
matrix.2011.red <- matrix.2011[rownames(matrix.2011) %in% genes,]
matrix.2012.red <- matrix.2012[rownames(matrix.2012) %in% genes,]
```
```{r checkDim}
print(dim(matrix.2011.red))
print(dim(matrix.2012.red))
print(dim(meta.2011.red))
print(dim(meta.2012.red))
```

```{r}
matrix <- cbind(matrix.2011.red, matrix.2012.red)

```

```{r}
ddsMatrix <- DESeqDataSetFromMatrix(matrix, colData = meta, design=~sample)
colData(ddsMatrix)$year <- factor(colData(ddsMatrix)$year)
```
# QA
## PCA
We use blind vst to plot a PCA and check for patterns and outliers
```{r}
vsd.QA <- vst(ddsMatrix, blind=TRUE)
geneNumber <- length(rownames(assay(vsd.QA)))
```

```{r}
plotPCA(vsd.QA, intgroup = c("treatment"), ntop=geneNumber)
```
```{r}
plotPCA(vsd.QA, intgroup = c("date"), ntop=geneNumber)
```

```{r}
plotPCA(vsd.QA, intgroup = c("year"), ntop=geneNumber)
```
```{r}
pca <- prcomp( t(assay(vsd.QA)))
percents <- round(summary(pca)$importance[2,]*100)

# standard colors for the paper
colors <- c("#e6e9bd","#b4d7bc","#7dc6bb","#1cb7ba","#00a7b9","#008b9a",
            "#849fba","#a181a6","#c76b97","#ee2e82","#8e5766","#8a7b73",
            "#869b7f","#42885e","#1aa64a","#75b443","#9bbe3b","#e1d51d",
            "#ffe200"
            )
df3D <- data.frame(PC1 = pca$x[,1], PC2 = pca$x[,2], PC3 = pca$x[,3],
                   treatment = meta$treatment, 
                   week = meta$date,
                   year = meta$year,
                   treatYear = paste(meta$year, meta$treatment, sep="-")
                   )

```
```{r}
g <- ggplot(df3D, aes(x = PC1, y = PC2)) +
  coord_fixed() +   
  geom_point(size=15, aes(shape= treatYear, color=week)) + 
	theme(legend.position = "none",
	      legend.box = "vertical",
	      axis.line = element_line(colour = "black"),
            panel.grid.major = element_blank(),
            panel.grid.minor = element_blank(),
            panel.border = element_blank(),
            panel.background = element_blank()
	      ) + 
  scale_shape_manual(values=c(0,15,1,16)) + 
  scale_color_manual(values=colors) +
  xlab(paste0("PC1: ",percents[1],"% variance")) +
	ylab(paste0("PC2: ",percents[2],"% variance"))

print(g)
```

```{r eval=FALSE, include=FALSE}
# save it for the paper
# change size to 15 before exporting
png(here("results", paste0("separated_2011_2012_spruce_PCA_", Sys.Date(), ".png")), width = 1920, heigh=1080, pointsize = 300)
print(g)
dev.off()
```

## PCA with year effect removed
We use blind vst to plot a PCA and check for patterns and outliers
```{r}
vsd.fix <- vsd.QA
assay(vsd.fix) <- limma::removeBatchEffect(assay(vsd.fix), vsd.fix$year)
```

```{r}
geneNumber <- length(rownames(assay(vsd.fix)))
```

```{r}
plotPCA(vsd.fix, intgroup = c("date"), ntop=geneNumber)
```

```{r}
plotPCA(vsd.fix, intgroup = c("year"), ntop=geneNumber)
```

```{r}
plotPCA(vsd.fix, intgroup = c("treatment"), ntop=geneNumber)
```

Final PCA
```{r}
pca <- prcomp( t(assay(vsd.fix)))
percents <- round(summary(pca)$importance[2,]*100)

# standard colors for the paper
colors <- c("#e6e9bd","#b4d7bc","#7dc6bb","#1cb7ba","#00a7b9","#008b9a",
            "#849fba","#a181a6","#c76b97","#ee2e82","#8e5766","#8a7b73",
            "#869b7f","#42885e","#1aa64a","#75b443","#9bbe3b","#e1d51d",
            "#ffe200"
            )
df3D <- data.frame(PC1 = pca$x[,1], PC2 = pca$x[,2], PC3 = pca$x[,3],
                   treatment = meta$treatment, 
                   week = meta$date,
                   year = meta$year,
                   treatYear = paste(meta$year, meta$treatment, sep="-")
                   )

```
```{r}
g <- ggplot(df3D, aes(x = PC1, y = PC2)) +
  coord_fixed() +   
  geom_point(size=4, aes(shape= treatYear, color=week)) + 
	theme(legend.position = "none",
	      legend.box = "vertical",
	      axis.line = element_line(colour = "black"),
            panel.grid.major = element_blank(),
            panel.grid.minor = element_blank(),
            panel.border = element_blank(),
            panel.background = element_blank()
	      ) + 
  scale_shape_manual(values=c(0,15,1,16)) + 
  scale_color_manual(values=colors) +
  xlab(paste0("PC1: ",percents[1],"% variance")) +
	ylab(paste0("PC2: ",percents[2],"% variance"))

print(g)
```
```{r eval=FALSE, include=FALSE}
# save it for the paper
# change size to 15 before exporting
png(here("results", paste0("combined_2011_2012_spruce_PCA_", Sys.Date(), ".png")), width = 1920, heigh=1080, pointsize = 300)
print(g)
dev.off()
```

## 3D PCA
We represent the first 3 compoments as a 3D interactive object

```{r}
plot_ly(df3D, x = ~PC1, y = ~PC3, z = ~PC2, 
        color = ~week, colors = colors, symbol= ~treatment,symbols = c("circle","square","cross","x"),
        text = ~paste("Date:", week, '<br>Treatment:', treatment) ) %>%   
  add_markers() %>%
  layout(scene = list(xaxis = list(title = paste("PC1 (",percents[1],"%)",sep="")),
                      yaxis = list(title = paste("PC3 (",percents[3],"%)",sep="")),
                      zaxis = list(title = paste("PC2 (",percents[2],"%)",sep=""))))

```



# DE

Load precalculated DE results from both years
```{r}
load(file.path(de2011Folder, "allDEresults.RData"))
control.2011.res <- control.res
fertilised.2011.res <- fertilised.res
combined.2011.res <- combined.res

load(file.path(de2012Folder, "allDEresults.RData"))
control.2012.res <- control.res
fertilised.2012.res <- fertilised.res
combined.2012.res <- combinedFvsC.res
```

Using the filterDE function we apply the padj of 0.01 and an absolute log2fc of 0.5
```{r filterStage, results="hide", message=FALSE}
control.2011.res.filter <- lapply(control.2011.res, filterDE)
fertilised.2011.res.filter <- lapply(fertilised.2011.res, filterDE)
combined.2011.res.filter <- lapply(combined.2011.res, filterDE)

control.2012.res.filter <- lapply(control.2012.res, filterDE)
fertilised.2012.res.filter <- lapply(fertilised.2012.res, filterDE)
combined.2012.res.filter <- lapply(combined.2012.res, filterDE)
```

## Plot DE genes
We can plot the results as positive and negative bars to reflect the amount of genes up or down
regulated in each comparition 

### Control date+1 vs date plot
```{r plotControl, echo=FALSE, results='asis'}
plotUpAndDown(control.2011.res.filter, xlabel = "2011 comparitions")
plotUpAndDown(control.2012.res.filter, xlabel = "2012 comparitions")
```

### Fertilised date+1 vs date plot
```{r plotFert}
plotUpAndDown(fertilised.2011.res.filter, xlabel = "2011 comparitions")
plotUpAndDown(fertilised.2012.res.filter, xlabel = "2012 comparitions")
```

### Fertilised vs Control per date plot
```{r plotCombined}
plotUpAndDown(combined.2011.res.filter, xlabel = "2011 comparitions")
plotUpAndDown(combined.2012.res.filter, xlabel = "2012 comparitions")
```

# Merged Differential Expression common timepoints
The dataset is ready for DESeq2. We check the design of the DE object 

```{r}
design(ddsMatrix)
```

We are interested on future comparitions like DE genes in each timepoints or between timepoints
that is why we created the interaction group

## Run DESeq
```{r runDE, eval=FALSE}
dds <- DESeq(ddsMatrix)
```

We check the result names
```{r}
resultsNames(dds)
```

## DE between years in common conditions and common timepoints

The common timepoint are weeks: 23, 25, 32 and 41
In order to make all the comparitions automated, we prepare vectors with the elements that will be used
in the result extraction, we generate 4 vectors, one per condition + year and we narrow it to the timepoints
of interest
```{r}
commonTimepoints <- c("23", "25", "32", "41")

control2011Vector <- as.character(unique(meta.2011.red$sample[grep("Control", meta.2011.red$sample)]))
control2011Vector <- sapply(commonTimepoints, function(x){control2011Vector[grep(x, control2011Vector)]})

control2012Vector <- as.character(unique(meta.2012.red$sample[grep("Control", meta.2012.red$sample)]))
control2012Vector <- sapply(commonTimepoints, function(x){control2012Vector[grep(x, control2012Vector)]})

fertilised2011Vector <- as.character(unique(meta.2011.red$sample[grep("Fertilised", meta.2011.red$sample)]))
fertilised2011Vector <- sapply(commonTimepoints, function(x){fertilised2011Vector[grep(x, fertilised2011Vector)]})

fertilised2012Vector <- as.character(unique(meta.2012.red$sample[grep("Fertilised", meta.2012.red$sample)]))
fertilised2012Vector <- sapply(commonTimepoints, function(x){fertilised2012Vector[grep(x, fertilised2012Vector)]})
```

## Result extraction


### Control date vs date
For this comparition we take consider only the control condition and we compare each pair of weeks
to determine which genes are up or down regaluated from one week to the previous week

```{r message=TRUE, warning=TRUE, include=FALSE}
load(here("env", "allDESpruceMergedResults.RData"))

```

```{r eval=FALSE}
control.res <- mapply(getRes, control2012Vector,
                      control2011Vector, 
                      MoreArgs = list(localDDS=dds, group="sample"))

fertilised.res <- mapply(getRes, fertilised2012Vector,
                      fertilised2011Vector, 
                      MoreArgs = list(localDDS=dds, group="sample"))


```

We add the corresponding names to the results
```{r eval=FALSE}
names(control.res) <- c("2012_Control_23 vs 2011_Control_23", 
                        "2012_Control_25 vs 2011_Control_25", 
                        "2012_Control_32 vs 2011_Control_32",
                        "2012_Control_41 vs 2011_Control_41")
names(fertilised.res) <- c("2012_Fertilised_23 vs 2011_Fertilised_23", 
                        "2012_Fertilised_25 vs 2011_Fertilised_25", 
                        "2012_Fertilised_32 vs 2011_Fertilised_32",
                        "2012_Fertilised_41 vs 2011_Fertilised_41")

```

The results contain all the genes, we need to filter them to get only those genes
that are stadistical significant and that have a strong condtition effect
For that we discard all the genes that has a padj higher than 0.01 and
a log2FoldChange lower than 0.5 (absolute value)
```{r filtering, results="hide", message=FALSE}
control.res.filter <- lapply(control.res, filterDE)
fertilised.res.filter <- lapply(fertilised.res, filterDE)
```

## Plot DE genes
We can plot the results as positive and negative bars to reflect the amount of genes up or down
regulated in each comparition 

### Control 2012 vs 2011 plot
```{r}
plotUpAndDown(control.res.filter)
```
### Fertilised 2012 vs 2011 plot
```{r}
plotUpAndDown(fertilised.res.filter)
```

## DE between Fertilised and Control individual years

We subset the combined results (Fertilised vs control) so both datasets have the same sampling point
```{r plotCombinedRed}
combined.2011.res.filter.red <- combined.2011.res.filter[c("23Fvs23C", "25Fvs25C", "32Fvs32C", "41Fvs41C")]
combined.2012.res.filter.red <- combined.2012.res.filter
names(combined.2012.res.filter.red) <- c("23Fvs23C", "25Fvs25C", "32Fvs32C", "41Fvs41C")
```

Plot to see the DE filtered genes
```{r}
plotUpAndDown(combined.2011.res.filter.red, xlabel = "2011 comparitions")
plotUpAndDown(combined.2012.res.filter.red, xlabel = "2012 comparitions")
```

Generate Venn diagrams to see the overlap between up and down regulated genes within each timepoint.
```{r}
timepoints <- c("23Fvs23C", "25Fvs25C", "32Fvs32C", "41Fvs41C")

sapply(timepoints, function(timepoint) {
  up2011 <- rownames(combined.2011.res.filter.red[[timepoint]][combined.2011.res.filter.red[[timepoint]]$log2FoldChange>0,])
  down2011 <- rownames(combined.2011.res.filter.red[[timepoint]][combined.2011.res.filter.red[[timepoint]]$log2FoldChange<0,])
  
  up2012 <- rownames(combined.2012.res.filter.red[[timepoint]][combined.2012.res.filter.red[[timepoint]]$log2FoldChange>0,])
  down2012 <- rownames(combined.2012.res.filter.red[[timepoint]][combined.2012.res.filter.red[[timepoint]]$log2FoldChange<0,])
  
  g <- draw.quad.venn(area1 = length(up2011), 
                        area2 = length(down2011), 
                        area3 = length(up2012),
                        area4 = length(down2012),
                        n12 = length(intersect(up2011, down2011)), 
                        n13 = length(intersect(up2011, up2012)),
                        n14 = length(intersect(up2011, down2012)),
                        n23 = length(intersect(down2011, up2012)), 
                        n24=length(intersect(down2011, down2012)),
                        n34=length(intersect(up2012, down2012)),
                        n123=length(intersect(up2011, intersect(down2011, up2012))), 
                        n124=length(intersect(up2011, intersect(down2011, down2012))), 
                        n134=length(intersect(up2011, intersect(up2012, down2012))), 
                        n234=length(intersect(down2011, intersect(up2012, down2012))),
                        n1234=length(intersect(intersect(up2011, down2011),intersect(up2012, down2012))),
                          category = c("Up 2011", "Down 2011", "Up 2012", "Down 2012"),
                         fill = c("orange", "red", "green", "blue"),
                          euler.d = TRUE, scaled = FALSE,
               cat.col = c("orange", "red", "green", "blue"),
               ind=F,
                          filename = NULL
                          )


  grid.arrange(gTree(children=g), top="Up & down regulated DE genes", bottom=paste("Comparition", timepoint))

  ### Averaging the log2foldchange of significant DE genes

  lfcList.red <- list(up2011 = combined.2011.res.filter.red[[timepoint]][combined.2011.res.filter.red[[timepoint]]$log2FoldChange>0,]$log2FoldChange,
                      up2012 = combined.2012.res.filter.red[[timepoint]][combined.2012.res.filter.red[[timepoint]]$log2FoldChange>0,]$log2FoldChange,
                      down2011 = combined.2011.res.filter.red[[timepoint]][combined.2011.res.filter.red[[timepoint]]$log2FoldChange<0,]$log2FoldChange,
                      down2012 = combined.2012.res.filter.red[[timepoint]][combined.2012.res.filter.red[[timepoint]]$log2FoldChange<0,]$log2FoldChange )
  
  boxplot(lfcList.red, ylim=c(-5,5), col=c("#00FF00", "#FF0000"), 
          ylab="log2FoldChange",
          main = paste("lfc distribution for: ", timepoint))


})

```






# Merged Differential Expression year effect
We need to change the design of the DE object to focus only on year

```{r}
design(ddsMatrix) <- ~year
```

We are interested on future comparitions like DE genes in each timepoints or between timepoints
that is why we created the interaction group

## Run DESeq
```{r runDEYear, eval=FALSE}
ddsYear <- DESeq(ddsMatrix)
```

We check the result names
```{r}
resultsNames(ddsYear)
```
Obtain the results
```{r evalYear}
year.res <- list(results(ddsYear ,
        name= "year_2012_vs_2011",
        filter = rowMedians(counts(ddsYear )),
        parallel = FALSE))
names(year.res) <- "2012vs2011"
```

Filter the results
```{r filterYear, results="hide", message=FALSE}
year.res.filter <- lapply(year.res, filterDE)
```

### 2012 vs 2011 plot
```{r}
plotUpAndDown(year.res.filter)
```

```{r saveresults,eval=FALSE, echo=FALSE}
save(dds, ddsYear, control.res, fertilised.res, year.res, 
     file=here("env", "allDESpruceMergedResults.RData"))
```

## DE between Fertilised and Control comparitions in each dataset

```{r}
load(file.path(de2011Folder, "ddsTreatment.RData"))
dds2011SpruceTreatment <- ddsSpruceTreatment
res2011SpruceTreatment <- resSpruceTreatment
res2011SpruceTreatment.filter <- lapply(res2011SpruceTreatment, filterDE)

load(file.path(de2012Folder, "ddsTreatment.RData"))
dds2012SpruceTreatment <- ddsSpruceTreatment
res2012SpruceTreatment <- resSpruceTreatment
res2012SpruceTreatment.filter <- lapply(res2012SpruceTreatment, filterDE)
```

2011 Treatment vs Control filtered plot
```{r}
plotUpAndDown(res2011SpruceTreatment.filter)
```

2012 Treatment vs Control filtered plot
```{r}
plotUpAndDown(res2012SpruceTreatment.filter)
```

```{r}
spruce2011Up <- rownames(res2011SpruceTreatment.filter$FertilisedvsControl[res2011SpruceTreatment.filter$FertilisedvsControl$log2FoldChange>0,])
spruce2011Down <- rownames(res2011SpruceTreatment.filter$FertilisedvsControl[res2011SpruceTreatment.filter$FertilisedvsControl$log2FoldChange<0,])

spruce2012Up <- rownames(res2012SpruceTreatment.filter$FertilisedvsControl[res2012SpruceTreatment.filter$FertilisedvsControl$log2FoldChange>0,])
spruce2012Down <- rownames(res2012SpruceTreatment.filter$FertilisedvsControl[res2012SpruceTreatment.filter$FertilisedvsControl$log2FoldChange<0,])

spruceUp <- intersect(spruce2011Up, spruce2012Up)
spruceDown <- intersect(spruce2011Down, spruce2012Down)
```

Up-regulated genes in common in both years
```{r}
grid.newpage()
draw.pairwise.venn(length(spruce2011Up),
                   length(spruce2012Up), 
                   length(spruceUp),
                          category = c("2011", "2012"), lty = rep("blank", 
    2), fill = c("orange", "red"), alpha = rep(0.5, 2), cat.pos = c(0, 
    0), cat.dist = rep(0.025, 2), scaled = T)
```

Down-regulated genes in common in both years
```{r}
grid.newpage()
draw.pairwise.venn(length(spruce2011Down),
                   length(spruce2012Down), 
                   length(spruceDown),
                          category = c("2011", "2012"), lty = rep("blank", 
    2), fill = c("orange", "red"), alpha = rep(0.5, 2), cat.pos = c(0, 
    0), cat.dist = rep(0.025, 2), scaled = T)
```

### Averaging the log2foldchange of significant DE genes
```{r}
lfcList <- list(Spruce11Up = res2011SpruceTreatment.filter$FertilisedvsControl[res2011SpruceTreatment.filter$FertilisedvsControl$log2FoldChange>0,]$log2FoldChange,
                Spruce12Up = res2012SpruceTreatment.filter$FertilisedvsControl[res2012SpruceTreatment.filter$FertilisedvsControl$log2FoldChange>0,]$log2FoldChange,
                Spruce11Down = res2011SpruceTreatment.filter$FertilisedvsControl[res2011SpruceTreatment.filter$FertilisedvsControl$log2FoldChange<0,]$log2FoldChange,
                Spruce12Down = res2012SpruceTreatment.filter$FertilisedvsControl[res2012SpruceTreatment.filter$FertilisedvsControl$log2FoldChange<0,]$log2FoldChange )

boxplot(lfcList, ylim=c(-5,5), col=c("#00FF00", "#FF0000"), ylab="log2FoldChange" )
```



# Correlation between similar timepoints and conditions across years

Calculate the mean of the replicates for the timepoints and comparitions that both years have in common
```{r}
matrix.2011.red <- data.frame(matrix.2011.red, check.names=F)
#colnames(matrix.2011.red) <- gsub("X","",colnames(matrix.2011.red))
matrix.2011.mean <- sapply(unique(colnames(matrix.2011.red)), function(x) { rowMeans(matrix.2011.red[,colnames(matrix.2011.red) %in% x])})
matrix.2011.mean <- floor(matrix.2011.mean)

matrix.2011.mean <- matrix.2011.mean[,c("2011_Fertilised_23", "2011_Fertilised_25", "2011_Fertilised_32", "2011_Fertilised_41", "2011_Control_23",
                      "2011_Control_25", "2011_Control_32", "2011_Control_41")]

matrix.2011.mean <-data.frame(matrix.2011.mean, stringsAsFactors = F)
colnames(matrix.2011.mean) <- gsub("X2011_","",colnames(matrix.2011.mean))

matrix.2012.red <- data.frame(matrix.2012.red, check.names=F)
matrix.2012.mean <- sapply(unique(colnames(matrix.2012.red)), function(x)  rowMeans(matrix.2012.red[colnames(matrix.2012.red) %in% x]))
matrix.2012.mean <- floor(matrix.2012.mean)
matrix.2012.mean <-data.frame(matrix.2012.mean, stringsAsFactors = F)
colnames(matrix.2012.mean) <- gsub("X2012_","",colnames(matrix.2012.mean))

```
Prepare for plot
```{r}
comparitions <- sort(colnames(matrix.2011.mean))

panel.cor <- function(x, y){
    usr <- par("usr"); on.exit(par(usr))
    par(usr = c(0, 1, 0, 1))
    r <- round(cor(x, y), digits=2)
    txt <- paste0("R = ", r)
    cex.cor <- 0.8/strwidth(txt)
    text(0.5, 0.5, txt, cex = cex.cor * r)
}
```

Correlation plots
```{r}
pairs(data.frame(matrix.2011.mean$Control_23, matrix.2012.mean$Control_23),  lower.panel = panel.cor)
pairs(data.frame(matrix.2011.mean$Fertilised_23, matrix.2012.mean$Fertilised_23),  lower.panel = panel.cor)
pairs(data.frame(matrix.2011.mean$Control_25, matrix.2012.mean$Control_25),  lower.panel = panel.cor)
pairs(data.frame(matrix.2011.mean$Fertilised_25, matrix.2012.mean$Fertilised_25),lower.panel = panel.cor)
pairs(data.frame(matrix.2011.mean$Control_32, matrix.2012.mean$Control_32),  lower.panel = panel.cor)
pairs(data.frame(matrix.2011.mean$Fertilised_32, matrix.2012.mean$Fertilised_32),  lower.panel = panel.cor)
pairs(data.frame(matrix.2011.mean$Control_41, matrix.2012.mean$Control_41),  lower.panel = panel.cor)
pairs(data.frame(matrix.2011.mean$Fertilised_41, matrix.2012.mean$Fertilised_41),  lower.panel = panel.cor)
```

Detected an outlier gene, we use an interactive version of the correlation plot to locate the gen
```{r}
highlight_key(data.frame(matrix.2011.mean$Control_23, matrix.2012.mean$Control_23)) %>%
  GGally::ggpairs(columns = 1:2) %>%
  ggplotly() %>%
  highlight("plotly_selected")
```

Gene index 27437, that corresponds to MA_171883g0010. According to Congenie this gene is related to TMV resistance N-like. 
Following the Uniprot link https://www.uniprot.org/uniprot/Q40392, its function is described as: Disease resistance protein. Resistance proteins guard the plant against pathogens that contain an appropriate avirulence protein via a direct or indirect interaction with this avirulence protein. That triggers a defense system including the hypersensitive response, which restricts the pathogen growth.

The gene has a high presence in the 2011 data but almost none in 2012 data. 

Since is only that gene that is an obvious outlier, wer remove it from both datasets
```{r}
matrix.2011.mean <- matrix.2011.mean[-27437,]
matrix.2012.mean <- matrix.2012.mean[-27437,]
```

Correlation plot, timepoint 23
```{r}
pairs(data.frame(matrix.2011.mean$Control_23, matrix.2012.mean$Control_23),  lower.panel = panel.cor)
pairs(data.frame(matrix.2011.mean$Fertilised_23, matrix.2012.mean$Fertilised_23),  lower.panel = panel.cor)
```
Correlation plot, timepoint 25
```{r}
pairs(data.frame(matrix.2011.mean$Control_25, matrix.2012.mean$Control_25),  lower.panel = panel.cor)
pairs(data.frame(matrix.2011.mean$Fertilised_25, matrix.2012.mean$Fertilised_25),lower.panel = panel.cor)
```
Correlation plot, timepoint 32
```{r}
pairs(data.frame(matrix.2011.mean$Control_32, matrix.2012.mean$Control_32),  lower.panel = panel.cor)
pairs(data.frame(matrix.2011.mean$Fertilised_32, matrix.2012.mean$Fertilised_32),  lower.panel = panel.cor)
```
Correlation plot, timepoint 41
```{r}
pairs(data.frame(matrix.2011.mean$Control_41, matrix.2012.mean$Control_41),  lower.panel = panel.cor)
pairs(data.frame(matrix.2011.mean$Fertilised_41, matrix.2012.mean$Fertilised_41),  lower.panel = panel.cor)
```

# Session information
```{r}
sessionInfo()
```