---
title: "Spruce roots Flakaliden project - Comparative year analysis"
author: "Alonso Serrano"
date: "`r Sys.Date()`"
output:
 html_document:
   toc: true
   number_sections: true
---

# Introduction
This file contains the script and results for the comparatived analysis done to the spruce roots Flakaliden 2011 & 2012 data. 

*200331 Creation

## Prerequisites
All RNAseq Fastq files where previously processed by a nextflow pipeline for RNAseq analysis.
The files where QA with fastQC, then trimmed by Trimmomatic and sortmeRNA and finally aligned by
Salmon.
The output of the pipeline are the Salmon files, one per original fastq file.

Metadata files containing all the information related to each sample, file name, time, condition and replica number.

Vst and DESeqMAtrix objects from 2011 and 2012 spruce data.

# Setup
Load libraries
```{r, message=FALSE}
library(DESeq2)
library(tximport)
library(BiocParallel)
library(plotly)
library(dplyr)
library(here)
library(limma)
register(MulticoreParam(4))
source(here("Rtoolbox/src/utilsDE.r"))
source(here("Rtoolbox/src/plot3Dvector.R"))
source(here("Rtoolbox/src/plotVectorPCA.R"))
source(here("Rtoolbox/src/plotUpAndDown.R"))
```

Set the project folder

```{r}
project2011Folder <- "/mnt/picea/projects/spruce/vhurry/14_SpruceRoots_Project"
de2011Folder <- file.path(project2011Folder, "DE")

project2012Folder <- "/mnt/picea/projects/spruce/vhurry/spruce-flakaliden-root-2012"
de2012Folder <- file.path(project2012Folder, "DE")
```

Load yearly datasets
```{r loadVSTpreprocessed, echo=FALSE}
load(file.path(de2011Folder, "vsd_QA.RData"))
vsd.QA.2011 <- vsd.QA
meta.2011 <- meta
ddsMatrix.2011 <- ddsMatrix

load(file.path(de2012Folder, "vsd_QA.RData"))
vsd.QA.2012 <- vsd.QA
meta.2012 <- meta
ddsMatrix.2012 <- ddsMatrix
```

## Dataset merge

Adapt 2011 meta
```{r}
meta.2011.red <- data.frame(id=meta.2011$Sample,
                            treatment=as.character(meta.2011$Treatment),
                            date=as.character(meta.2011$Week),
                            year=rep("2011", length(meta.2011$Sample)),
                            stringsAsFactors = F)
meta.2011.red$treatment[meta.2011.red$treatment == 'C'] <- 'Control'
meta.2011.red$treatment[meta.2011.red$treatment == 'F'] <- 'Fertilised'
meta.2011.red$date <- gsub("W","",meta.2011.red$date)
meta.2011.red$sample <- paste("2011", meta.2011.red$treatment, 
                              meta.2011.red$date, sep = "_")
```
Adapt 2012 meta
```{r}
meta.2012.red <- data.frame(id=meta.2012$SciLifeID,
                            treatment=as.character(meta.2012$treatment),
                            date=as.character(meta.2012$date),
                             year=rep("2012", length(meta.2012$SciLifeID)),
                            stringsAsFactors = F)

meta.2012.red$date[meta.2012.red$date == 'Early_June'] <- '23'
meta.2012.red$date[meta.2012.red$date == 'Late_June'] <- '25'
meta.2012.red$date[meta.2012.red$date == 'August'] <- '32'
meta.2012.red$date[meta.2012.red$date == 'October'] <- '41'

meta.2012.red$sample <- paste("2012", meta.2012.red$treatment, 
                              meta.2012.red$date, sep = "_")
```
Merge metadata
```{r mergeMeta}
meta <- rbind(meta.2011.red, meta.2012.red)
```

Expression data
```{r}
matrix.2011 <- assay(ddsMatrix.2011)
all(colnames(matrix.2011) == meta.2011.red$id)
colnames(matrix.2011) <- meta.2011.red$sample

```

```{r}
matrix.2012 <- assay(ddsMatrix.2012)
all(colnames(matrix.2012) == meta.2012.red$id)
colnames(matrix.2012) <- meta.2012.red$sample
```

Genes in common
```{r}
genes <- intersect(rownames(ddsMatrix.2011), rownames(ddsMatrix.2012))

genes <- sort(genes)
```

```{r}
matrix.2011.red <- matrix.2011[rownames(matrix.2011) %in% genes,]
matrix.2012.red <- matrix.2012[rownames(matrix.2012) %in% genes,]
```
```{r checkDim}
print(dim(matrix.2011.red))
print(dim(matrix.2012.red))
print(dim(meta.2011.red))
print(dim(meta.2012.red))
```

```{r}
matrix <- cbind(matrix.2011.red, matrix.2012.red)
```


```{r}
ddsMatrix <- DESeqDataSetFromMatrix(matrix, colData = meta, design=~sample)
```
# QA
## PCA
We use blind vst to plot a PCA and check for patterns and outliers

```{r}
vsd.QA <- vst(ddsMatrix, blind=TRUE)
geneNumber <- length(rownames(assay(vsd.QA)))
```

```{r}
plotPCA(vsd.QA, intgroup = c("treatment"), ntop=geneNumber)
```
```{r}
plotPCA(vsd.QA, intgroup = c("date"), ntop=geneNumber)
```

```{r}
plotPCA(vsd.QA, intgroup = c("year"), ntop=geneNumber)
```


```{r}
vsd.fix <- vsd.QA
assay(vsd.fix) <- limma::removeBatchEffect(assay(vsd.fix), vsd.fix$year)
```
# QA
## PCA
We use blind vst to plot a PCA and check for patterns and outliers

```{r}
geneNumber <- length(rownames(assay(vsd.QA.fix)))
```

```{r}
plotPCA(vsd.fix, intgroup = c("date"), ntop=geneNumber)
```

```{r}
plotPCA(vsd.fix, intgroup = c("year"), ntop=geneNumber)
```

```{r}
plotPCA(vsd.fix, intgroup = c("treatment"), ntop=geneNumber)
```
