---
title: "Flakaliden spruce roots + fungi cluster analysis"
author: "Alonso Serrano"
date: "`r format(Sys.time(), '%d %B %Y')`"
output: 
  html_document: 
    number_sections: yes
    toc: yes
---


```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```
# Introduction
This file contains the script and results of the data from spruce roots and fungi 
from the 2011 Flakaliden dataset. 

* 20191206 Creation  

## Prerequisites
The variance stabilization transformed dataset, either as TSV or RData obtained from de differential expression analysis step.
Infomap clustering results from Seidr network

## Setup
```{r generalSetup}
library(dplyr)
library(here)
source("~/Git/Rtoolbox/src/plotEigenGene.R")
source("~/Git/Rtoolbox/src/getEigengenes.R")
source("~/Git/UPSCb-common/src/R/gopher.R")
conditions <- c("control", "fertilised", "combined")

spruceDir <- "/mnt/picea/projects/spruce/vhurry/14_SpruceRoots_Project"
fungiDir <- "/mnt/picea/projects/spruce/vhurry/fungi-flakaliden-2011"

spruceDE <- file.path(spruceDir,"DE")
fungiDE <- file.path(fungiDir,"DE")

checkFile <- function(fileName) {
  if (!file.exists(fileName)) stop(paste("File", fileName, "doesn't exist"))
}
```

## Check files

### VSD data file
The files are tab separated, rows contain sample meta information (week and treatment),
columns contain gene or KEGG orthologs names. The values are expression data after variance stabilization transformation

### Cluster files
They must contain two columns gene and cluster and be tab separated<br />
ex.<br />
gene	cluster<br />
MA_5320g0010	Cluster1<br />
MA_32763g0010	Cluster1<br />
MA_66378g0010	Cluster1<br />


```{r checkFiles, eval=FALSE}
# load spruce cluster files
controlSpruceDataFile <- file.path(spruceDE, "controlVsdData.tsv")
fertilisedSpruceDataFile <- file.path(spruceDE, "fertilisedVsdData.tsv")
combinedSpruceDataFile <- file.path(spruceDE, "combinedVsdData.tsv")
checkFile(controlSpruceDataFile)
checkFile(fertilisedSpruceDataFile)
checkFile(combinedSpruceDataFile)

# load fungi cluster files
controlFungiDataFile <- file.path(fungiDE, "controlVsdData.tsv")
fertilisedFungiDataFile <- file.path(fungiDE, "fertilisedVsdData.tsv")
combinedFungiDataFile <- file.path(fungiDE, "combinedVsdData.tsv")
checkFile(controlFungiDataFile)
checkFile(fertilisedFungiDataFile)
checkFile(combinedFungiDataFile)
```

```{r loadRdata}
load(file.path(spruceDE, "controlVsdData.RData"))
controlSpruceData <- controlData
load(file.path(spruceDE, "fertilisedVsdData.RData"))
fertilisedSpruceData <- fertilisedData
load(file.path(spruceDE, "combinedVsdData.RData"))
combinedSpruceData <- combinedData

load(file.path(fungiDE, "controlVsdData.RData"))
controlFungiData <- controlData
load(file.path(fungiDE, "fertilisedVsdData.RData"))
fertilisedFungiData <- fertilisedData
load(file.path(fungiDE, "combinedVsdData.RData"))
combinedFungiData <- combinedData

```

```{r}
# load spruce data files
controlSpruceClusterFile <- file.path(spruceDir,"spruce-networks/control/cluster","InfomapClusters.tsv")
fertilisedSpruceClusterFile <- file.path(spruceDir,"spruce-networks/fertilised/cluster","InfomapClusters.tsv")
combinedSpruceClusterFile <- file.path(spruceDir,"spruce-networks/combined/cluster","InfomapClusters.tsv")

checkFile(controlSpruceClusterFile)
checkFile(fertilisedSpruceClusterFile)
checkFile(combinedSpruceClusterFile)

# load fungi data files
controlFungiClusterFile <- file.path(fungiDir,"networks/control/cluster","InfomapClusters.tsv")
fertilisedFungiClusterFile <- file.path(fungiDir,"networks/fertilised/cluster","InfomapClusters.tsv")
combinedFungiClusterFile <- file.path(fungiDir,"networks/combined/cluster","InfomapClusters.tsv")

checkFile(controlFungiClusterFile)
checkFile(fertilisedFungiClusterFile)
checkFile(combinedFungiClusterFile)

```

```{r include=FALSE}
#load(here("temp","191120.RData"))
```

## Load files
```{r loadFiles, eval=FALSE}
# controlSpruceData <- read.table(controlSpruceDataFile, stringsAsFactors=F, header=T, sep='\t', row.names=1, comment.char="")
# fertilisedSpruceData <- read.table(fertilisedSpruceDataFile, stringsAsFactors=F, header=T, sep='\t', row.names=1, comment.char="")
# combinedSpruceData <- read.table(combinedSpruceDataFile, stringsAsFactors=F, header=T, sep='\t', row.names=1, comment.char="")

#name correction, spruce rownames, have an extra dash
rownames(controlSpruceData) <- gsub("-", "", rownames(controlSpruceData))
rownames(fertilisedSpruceData) <- gsub("-", "", rownames(fertilisedSpruceData))
rownames(combinedSpruceData) <- gsub("-", "", rownames(combinedSpruceData))

# controlFungiData <- read.table(controlFungiDataFile, stringsAsFactors=F, header=T, sep='\t', row.names=1, comment.char="")
# fertilisedFungiData <- read.table(fertilisedFungiDataFile, stringsAsFactors=F, header=T, sep='\t', row.names=1, comment.char="")
# combinedFungiData <- read.table(combinedFungiDataFile, stringsAsFactors=F, header=T, sep='\t', row.names=1, comment.char="")

controlSpruceClusters <- read.table(controlSpruceClusterFile, stringsAsFactors=F, header=T, sep='\t', comment.char="")
fertilisedSpruceClusters <- read.table(fertilisedSpruceClusterFile, stringsAsFactors=F, header=T, sep='\t', comment.char="")
combinedSpruceClusters <- read.table(combinedSpruceClusterFile, stringsAsFactors=F, header=T, sep='\t', comment.char="")

controlFungiClusters <- read.table(controlFungiClusterFile, stringsAsFactors=F, header=T, sep='\t', comment.char="")
fertilisedFungiClusters <- read.table(fertilisedFungiClusterFile, stringsAsFactors=F, header=T, sep='\t', comment.char="")
combinedFungiClusters <- read.table(combinedFungiClusterFile, stringsAsFactors=F, header=T, sep='\t', comment.char="")
```

## Create metadata based on the information on sample names
```{r}
controlSpruceGroup <- substr(rownames(controlSpruceData),4,4)
controlSpruceTime <- as.integer(substr(rownames(controlSpruceData),2,3))

fertilisedSpruceGroup <- substr(rownames(fertilisedSpruceData),4,4)
fertilisedSpruceTime <- as.integer(substr(rownames(fertilisedSpruceData),2,3))

combinedSpruceGroup <- substr(rownames(combinedSpruceData),4,4)
combinedSpruceTime <- as.integer(substr(rownames(combinedSpruceData),2,3))

controlFungiGroup <- substr(rownames(controlFungiData),4,4)
controlFungiTime <- as.integer(substr(rownames(controlFungiData),2,3))

fertilisedFungiGroup <- substr(rownames(fertilisedFungiData),4,4)
fertilisedFungiTime <- as.integer(substr(rownames(fertilisedFungiData),2,3))

combinedFungiGroup <- substr(rownames(combinedFungiData),4,4)
combinedFungiTime <- as.integer(substr(rownames(combinedFungiData),2,3))
```

# Profile matching function
```{r profilefunction}
profileSimilarity <- function(data1, data2, clusterData1, clusterData2, 
                              group1, group2, time1, time2, name1, name2){

  #resultados
  # cluster1F -> cluster1S -> 0.34
  clusters1 <- unique(clusterData1$cluster)
  clusters2 <- unique(clusterData2$cluster)
  
  res <- data.frame(name1 = character(),
                   name2 = character(), 
                   score = double(), 
                   stringsAsFactors = FALSE) 

  #por cada cluster
  sapply(clusters1, function(clus1){
    
    clus1i <- paste0(clus1, "i")
    p1 <- plotEigengene(data1,
                clusterData1[clusterData1$cluster==clus1,]$gene, 
                group1, time1)
    p1.data <- p1$data %>%
      group_by(x) %>%
      summarise(n = n(), y = mean(y, na.rm = TRUE) )
    eigen1 <- p1.data$y
    eigen1i <- p1.data$y * -1
  
    sapply(clusters2, function(clus2){
      #print(paste("comparing",clus1, clus2))
      clus2i <- paste0(clus2, "i")
        p2 <- plotEigengene(data2,
              clusterData2[clusterData2$cluster==clus2,]$gene, 
              group2, time2)
        p2.data <- p2$data %>%
          group_by(x) %>%
          summarise(n = n(), y = mean(y, na.rm = TRUE) )
        eigen2 <- p2.data$y
        eigen2i <- p2.data$y * -1
        
        #add comparitions
        #1vs2
        res <<- rbind(res, c(clus1, clus2 ,cor(eigen1, eigen2, method = c("spearman"))), 
                   stringsAsFactors = FALSE)
        #1ivs2
        res <<- rbind(res, c(clus1i, clus2 ,cor(eigen1i, eigen2, method = c("spearman"))),
                   stringsAsFactors = FALSE)
        #1vs2i
        res <<- rbind(res, c(clus1, clus2i ,cor(eigen1, eigen2i, method = c("spearman"))),
                   stringsAsFactors = FALSE)
        #1ivs2i
        res <<- rbind(res, c(clus1i, clus2i ,cor(eigen1i, eigen2i, method = c("spearman"))),
                   stringsAsFactors = FALSE)
    })

  })
  # retorno
  names(res) <- c(name1, name2, "score")
  res$score <- as.numeric(res$score)
  return(res)
}
```


# Control profile matching
```{r controlMatching, echo=TRUE, message=FALSE}
control.res <- profileSimilarity(controlSpruceData, controlFungiData, controlSpruceClusters, controlFungiClusters,
                         controlSpruceGroup, controlFungiGroup, controlSpruceTime, controlFungiTime, 
                         "Spruce-control", "Fungi-control")

control.res.sort <- control.res[order(control.res$score, decreasing = T),]
```
## Best correlated
```{r controlHead}
head(control.res.sort)
```
## Best reverse correlated
```{r controlTail}
tail(control.res.sort)
```
## Save results to file
```{r controlSave}
# control.res.sort$`Spruce-control`<- paste0("spruce",control.res.sort$`Spruce-control`)
# control.res.sort$`Fungi-control`<- paste0("fungi",control.res.sort$`Fungi-control`)
write.table(control.res.sort, file=here("results",paste0("controlProfileMatching_",Sys.Date(),".tsv")), sep='\t', quote=F, row.names=F)
```


# Fertilised profile matching
```{r fertilisedMatching, echo=TRUE, message=FALSE}
fertilised.res <- profileSimilarity(fertilisedSpruceData, fertilisedFungiData, fertilisedSpruceClusters, fertilisedFungiClusters,
                         fertilisedSpruceGroup, fertilisedFungiGroup, fertilisedSpruceTime, fertilisedFungiTime, 
                         "Spruce-fertilised", "Fungi-fertilised")

fertilised.res.sort <- fertilised.res[order(fertilised.res$score, decreasing = T),]
```
## Best correlated
```{r fertilisedHead}
head(fertilised.res.sort)
```
## Best reverse correlated
```{r fertilisedTail}
tail(fertilised.res.sort)
```
## Save results to file
```{r fertilisedSave}
# fertilised.res.sort$`Spruce-fertilised`<- paste0("spruce",fertilised.res.sort$`Spruce-fertilised`)
# fertilised.res.sort$`Fungi-fertilised`<- paste0("fungi",fertilised.res.sort$`Fungi-fertilised`)
write.table(fertilised.res.sort, file=here("results", paste0("fertilisedProfileMatching_",Sys.Date(),".tsv") ), sep='\t', quote=F, row.names=F)
```

# Combined profile matching
```{r combinedMatching, echo=TRUE, message=FALSE}
combined.res <- profileSimilarity(combinedSpruceData, combinedFungiData, combinedSpruceClusters, combinedFungiClusters,
                         combinedSpruceGroup, combinedFungiGroup, combinedSpruceTime, combinedFungiTime, 
                         "Spruce-combined", "Fungi-combined")

combined.res.sort <- combined.res[order(combined.res$score, decreasing = T),]
```
## Best correlated
```{r combinedHead}
head(combined.res.sort)
```
## Best reverse correlated
```{r combinedTail}
tail(combined.res.sort)
```
## Save results to file
```{r combinedSave}
# combined.res.sort$`Spruce-combined`<- paste0("spruce",combined.res.sort$`Spruce-combined`)
# combined.res.sort$`Fungi-combined`<- paste0("fungi",combined.res.sort$`Fungi-combined`)
write.table(combined.res.sort, file=here("results", paste0("combinedProfileMatching_",Sys.Date(),".tsv")), sep='\t', quote=F, row.names=F)
```

```{r getClustersAmounts}
cluster2Network <- function(data, clusterData, species){
  
  temp <- lapply(unique(clusterData$cluster), function(localCluster) {
    spruceGenes <- getEigenGenes(data, clusterData[clusterData$cluster==localCluster,]$gene)
  })
  #spruceCluster <- "Cluster1"
  names(temp) <- paste0(species, unique(clusterData$cluster) )
      
  posTemp <- lapply(temp, function(x){
    x$positive
  })
  
  negTemp <- lapply(temp, function(x){
    x$negative
  })
  names(negTemp) <- paste0(names(negTemp),"i")
  
  posResult <- sapply(posTemp, function(x) {
    length(x)
  })
  negResult <- sapply(negTemp, function(x) {
    length(x)
  })
  
  posDF <- data.frame(clusters=names(posResult), geneNumber=posResult)
  negDF <- data.frame(clusters=names(negResult), geneNumber=negResult)
  res <- rbind(posDF, negDF)
  return (res)
}

spruceControl <- cluster2Network(controlSpruceData, controlSpruceClusters, "spruce")
fungiControl <- cluster2Network(controlFungiData, controlFungiClusters, "fungi")
write.table(rbind(spruceControl, fungiControl), file="controlClusterNetworkMeta.tsv", sep='\t', row.names = F, quote=F)

spruceFertilised <- cluster2Network(fertilisedSpruceData, fertilisedSpruceClusters, "spruce")
fungiFertilised <- cluster2Network(fertilisedFungiData, fertilisedFungiClusters, "fungi")
write.table(rbind(spruceFertilised, fungiFertilised), file="fertilisedClusterNetworkMeta.tsv", sep='\t', row.names = F, quote=F)
```





```{r miniFunction}
getPlotData <-function(data1, genes1, group1, time1) {
  p1 <- plotEigengene(data1, genes1, group1, time1)
    p1.data <- p1$data %>%
      group_by(x) %>%
      summarise(n = n(), y = mean(y, na.rm = TRUE) )
  return(p1.data)
}
```



```{r profileDegradationFunction}

getProfileDegradationMatrix <- function(pairedData, conditionSpruceClusters, conditionFungiClusters, 
                                  data1Spruce, data1Fungi,data2Spruce, data2Fungi,
                                  group1Spruce, group1Fungi,group2Spruce, group2Fungi,
                                  time1Spruce, time1Fungi,time2Spruce, time2Fungi,  
                                  
                                  conditionNames = c("control", "fertilised") ){

  condition.pos <- pairedData #[pairedData$score>0,]
  
  score1 <- vector(mode = "numeric", length = length(rownames(condition.pos)))
  score2 <- vector(mode = "numeric", length = length(rownames(condition.pos)))
  
  sapply(1:length(rownames(condition.pos)), function(l) {
    
   
    spruceCluster <- condition.pos[l,1]
    fungiCluster <- condition.pos[l,2]
    spruceReverse <- ifelse(endsWith(spruceCluster, "i"), TRUE, FALSE) 
    fungiReverse <- ifelse(endsWith(fungiCluster, "i"), TRUE, FALSE)
  
    spruceCluster <- ifelse(endsWith(spruceCluster, "i"), substr(spruceCluster, 1, nchar(spruceCluster)-1), spruceCluster) 
    fungiCluster <- ifelse(endsWith(fungiCluster, "i"), substr(fungiCluster, 1, nchar(fungiCluster)-1), fungiCluster) 
  
    c1SpruceClusterTotalGenes <- conditionSpruceClusters[conditionSpruceClusters$cluster==spruceCluster,]$gene
    c1FungiClusterTotalGenes <- conditionFungiClusters[conditionFungiClusters$cluster==fungiCluster,]$gene
  
    spruceGenes <- getEigenGenes(data1Spruce, c1SpruceClusterTotalGenes)
    c1SpruceClusterGenes <- if(spruceReverse) spruceGenes$negative else spruceGenes$positive
  
    fungiGenes <- getEigenGenes(data1Fungi, c1FungiClusterTotalGenes)
    c1FungiClusterGenes <- if(fungiReverse) fungiGenes$negative else fungiGenes$positive 
  
    # we extract only the genes that exist in both conditions and forget about uniques for the moment
    commonSpruceClusterGenes <- intersect(c1SpruceClusterGenes, colnames(data2Spruce))
    commonFungiClusterGenes <- intersect(c1FungiClusterGenes, colnames(data2Fungi))
    
    # TODO extract unique genes that exist in c1 but not in c2
  
    #initialise cor1 and cor2 for those cases that don't have genes in common
    cor1 <- 0
    cor2 <- 0
    if(length(commonSpruceClusterGenes)>0 && length(commonFungiClusterGenes)>0) {
      p1Spruce <- getPlotData(data1Spruce, commonSpruceClusterGenes, group1Spruce, time1Spruce)
      p1Fungi <- getPlotData(data1Fungi, commonFungiClusterGenes, group1Fungi, time1Fungi)
    
      p2Spruce <- getPlotData(data2Spruce, commonSpruceClusterGenes, group2Spruce, time2Spruce)
      p2Fungi <-getPlotData(data2Fungi, commonFungiClusterGenes, group2Fungi, time2Fungi)
      cor1 <- cor(p1Spruce$y, p1Fungi$y, method = c("spearman"))
      cor2 <- cor(p2Spruce$y, p2Fungi$y, method = c("spearman"))
    }
  
    score1[l] <<- cor1
    score2[l] <<- cor2
    
  }) #end of sapply

  res <- condition.pos[,1:2]
  res[[paste0(conditionNames[1],"_score")]] <- score1
  res[[paste0(conditionNames[2],"_score")]] <- score2
  res$degradation <- score1 - score2
  res.sort <- res[order(res$degradation, decreasing = T),]
  
  #write.table(fertilised2control.sort, file="fertilised2control.sort.tsv", quote = F, sep='\t', row.names = F)
  #
  res.sort

}
```
```{r controlDEF}
controlDEG <- getProfileDegradationMatrix(control.res.sort, 
                                          conditionSpruceClusters=controlSpruceClusters, 
                                          conditionFungiClusters=controlFungiClusters,
                                          data1Spruce=controlSpruceData, 
                                          data1Fungi=controlFungiData, 
                                          data2Spruce=fertilisedSpruceData, 
                                          data2Fungi=fertilisedFungiData,
                                          group1Spruce=controlSpruceGroup, 
                                          group1Fungi=controlFungiGroup,
                                          group2Spruce=fertilisedSpruceGroup , 
                                          group2Fungi=fertilisedFungiGroup ,
                                          time1Spruce=controlSpruceTime , 
                                          time1Fungi=controlFungiTime,
                                          time2Spruce=fertilisedSpruceTime, 
                                          time2Fungi=fertilisedFungiTime ,
                                          conditionNames = c("control", "fertilised")
                                          
                                          )


```

```{r controlViolin}

controlDEG.df <- controlDEG[controlDEG$control_score>0,]

controlDEG.df <- controlDEG.df[order(controlDEG.df$degradation, decreasing = F),]
controlDEG.df <- melt(controlDEG.df[,c("control_score","fertilised_score")])
 
ggplot(controlDEG.df, aes(x=variable, y=value  )) + 
  geom_violin() +
  stat_summary(fun=mean, geom="point", shape=23, size=2) + 
  stat_summary(fun=median, geom="point", size=2, color="red") +
  geom_boxplot(width=0.1)



write.table(controlDEG.df, file=here("results",paste0("controlDEG.df.sort_",Sys.Date(),".tsv") ) , quote = F, sep='\t', row.names = F)

```


```{r fertilisedDEG}
fertilisedDEG <- getProfileDegradationMatrix(fertilised.res.sort, 
                                          conditionSpruceClusters=fertilisedSpruceClusters, 
                                          conditionFungiClusters=fertilisedFungiClusters,
                                          data1Spruce=fertilisedSpruceData, 
                                          data1Fungi=fertilisedFungiData, 
                                          data2Spruce=controlSpruceData, 
                                          data2Fungi=controlFungiData,
                                          group1Spruce=fertilisedSpruceGroup, 
                                          group1Fungi=fertilisedFungiGroup,
                                          group2Spruce=controlSpruceGroup, 
                                          group2Fungi=controlFungiGroup,
                                          time1Spruce=fertilisedSpruceTime, 
                                          time1Fungi=fertilisedFungiTime,
                                          time2Spruce=controlSpruceTime, 
                                          time2Fungi=controlFungiTime,
                                          conditionNames = c("fertilised", "control")
                                          
                                          )


```

```{r fertilisedViolin}
 
 
fertilisedDEG.df <- fertilisedDEG[fertilisedDEG$fertilised_score>0,]

fertilisedDEG.df <- fertilisedDEG.df[order(fertilisedDEG.df$degradation, decreasing = F),]
fertilisedDEG.df <- melt(fertilisedDEG.df[,c("fertilised_score","control_score")])
 
p <- ggplot(fertilisedDEG.df, aes(x=variable, y=value  )) + geom_violin() +
   stat_summary(fun.y=median, geom="point", size=2, color="red")

write.table(fertilisedDEG.df, file=here("results",paste0("fertilisedDEG.df.sort_",Sys.Date(),".tsv") ), quote = F, sep='\t', row.names = F)

```
```{r}


length(control.res$score) #2816
length(fertilised.res$score) #2640
combined.df <- data.frame(condition=c(rep("control", length(control.res$score)),
                                      rep("fertilised", length(fertilised.res$score))),
                          data=c(control.res$score, fertilised.res$score)
                          )

p <- ggplot(combined.df, aes(x=condition, y=data)) + geom_violin(scale = "area") + geom_jitter(shape=16, position=position_jitter(0.2))

p <- ggplot(combined.df, aes(x=condition, y=data)) + geom_violin(scale = "area") +
  geom_dotplot(binaxis='y', stackdir='center', dotsize=0.1)
p <- ggplot(combined.df, aes(x=condition, y=data)) + geom_boxplot()
```


```{r}
controlAvg <- mean(controlDEG$degradation)
fertilisedAvg <- mean(fertilisedDEG$degradation)

print(controlAvg)
print(fertilisedAvg)


```



```{r}
# fertilised2control <- cbind(condition.pos[,1:2], score, scoreInControl)
# fertilised2control$degradation <- fertilised2control$score - fertilised2control$scoreInControl
# fertilised2control.sort <- fertilised2control[order(fertilised2control$degradation, decreasing = F),]
# 
# fertilised2control.pos <- fertilised2control[fertilised2control$score>0,]
# fertilised2control.pos.sort <- fertilised2control.pos[order(fertilised2control.pos$degradation, decreasing = F),]
# 
# write.table(fertilised2control.sort, file="fertilised2control.sort.tsv", quote = F, sep='\t', row.names = F)
```

```{r}
#keep positive relations
condition.pos <- control.res.sort[control.res.sort$score>0,]
#condition.pos <- control.res.sort
#otherCondition.pos <-  fertilised.res.sort[fertilised.res.sort$score>0,]

conditionSpruceClusters <- controlSpruceClusters
#otherConditionSpruceClusters <- fertilisedSpruceClusters
conditionFungiClusters <- controlFungiClusters
#otherConditionFungiClusters <- fertilisedFungiClusters

data1Spruce <- controlSpruceData
data1Fungi  <- controlFungiData
data2Spruce <- fertilisedSpruceData
data2Fungi  <- fertilisedFungiData

group1Spruce <- controlSpruceGroup 
group1Fungi  <- controlFungiGroup 
group2Spruce <- fertilisedSpruceGroup 
group2Fungi  <- fertilisedFungiGroup 

time1Spruce <- controlSpruceTime
time1Fungi  <- controlFungiTime 
time2Spruce <- fertilisedSpruceTime
time2Fungi  <- fertilisedFungiTime 


# control2fertilised <- sapply(1:length(rownames(condition.pos)), function(l) {
score <- vector(mode = "numeric", length = length(rownames(condition.pos)))
scoreInFertilised <- vector(mode = "numeric", length = length(rownames(condition.pos)))

sapply(1:length(rownames(condition.pos)), function(l) {
  spruceCluster <- condition.pos[l,1]
  fungiCluster <- condition.pos[l,2]
  spruceReverse <- ifelse(endsWith(spruceCluster, "i"), TRUE, FALSE) 
  fungiReverse <- ifelse(endsWith(fungiCluster, "i"), TRUE, FALSE)

  spruceCluster <- ifelse(endsWith(spruceCluster, "i"), substr(spruceCluster, 1, nchar(spruceCluster)-1), spruceCluster) 
  fungiCluster <- ifelse(endsWith(fungiCluster, "i"), substr(fungiCluster, 1, nchar(fungiCluster)-1), fungiCluster) 

  c1SpruceClusterTotalGenes <- conditionSpruceClusters[conditionSpruceClusters$cluster==spruceCluster,]$gene
  c1FungiClusterTotalGenes <- conditionFungiClusters[conditionFungiClusters$cluster==fungiCluster,]$gene

  spruceGenes <- getEigenGenes(data1Spruce, c1SpruceClusterTotalGenes)
  c1SpruceClusterGenes <- if(spruceReverse) spruceGenes$negative else spruceGenes$positive

  fungiGenes <- getEigenGenes(data1Fungi, c1FungiClusterTotalGenes)
  c1FungiClusterGenes <- if(fungiReverse) fungiGenes$negative else fungiGenes$positive 

  # we extract only the genes that exist in both conditions and forget about uniques for the moment
  commonSpruceClusterGenes <- intersect(c1SpruceClusterGenes, colnames(data2Spruce))
  commonFungiClusterGenes <- intersect(c1FungiClusterGenes, colnames(data2Fungi))
  
  # TODO extract unique genes that exist in c1 but not in c2

  #initialise cor1 and cor2 for those cases that don't have genes in common
  cor1 <- 0
  cor2 <- 0
  if(length(commonSpruceClusterGenes)>0 && length(commonFungiClusterGenes)>0) {
    p1Spruce <- getPlotData(data1Spruce, commonSpruceClusterGenes, group1Spruce, time1Spruce)
    p1Fungi <- getPlotData(data1Fungi, commonFungiClusterGenes, group1Fungi, time1Fungi)
  
    p2Spruce <- getPlotData(data2Spruce, commonSpruceClusterGenes, group2Spruce, time2Spruce)
    p2Fungi <-getPlotData(data2Fungi, commonFungiClusterGenes, group2Fungi, time2Fungi)
    cor1 <- cor(p1Spruce$y, p1Fungi$y, method = c("spearman"))
    cor2 <- cor(p2Spruce$y, p2Fungi$y, method = c("spearman"))
  }

  score[l] <<- cor1
  scoreInFertilised[l] <<- cor2
  
})

control2fertilised <- cbind(condition.pos[,1:2], score, scoreInFertilised)
control2fertilised$degradation <- control2fertilised$score - control2fertilised$scoreInFertilised
control2fertilised.sort <- control2fertilised[order(control2fertilised$degradation, decreasing = F),]
#
control2fertilised.pos <- control2fertilised[control2fertilised$score>0,]
control2fertilised.pos.sort <- control2fertilised.pos[order(control2fertilised.pos$degradation, decreasing = F),]

write.table(control2fertilised.sort, file="control2fertilised.sort.tsv", quote = F, sep='\t', row.names = F)
 
```


```{r}
library(reshape2)
d <- melt(control2fertilised.pos.sort[,c("score","scoreInFertilised")])
 
 library(ggplot2)
# Basic violin plot

p <- ggplot(d, aes(x=variable, y=value  )) + 
  #geom_violin()
  geom_violin()
p

```
Get the maximum difference.

```{r}
res.pos <- control2fertilised %>% 
  filter((score > 0) & (scoreInFertilised > 0)) #%>% 
  #filter((degradation<1) & (degradation > -1))


res.pos <- res.pos[order(res.pos$degradation, decreasing = T),]

res.neg <- control2fertilised %>% 
  filter((score < 0) & (scoreInFertilised < 0)) #%>% 
  #filter((degradation<1) & (degradation > -1))


res.neg <- res.neg[order(res.neg$degradation, decreasing = T),]
```
```{r}
mostDegraded <- rbind(head(res.pos, n=5),
tail(res.pos, n=5),
head(res.neg, n=5),
tail(res.neg, n=5))

```


```{r}


enrichThis <- function(data1, data2, clusterData1, group1, time1, group2, time2, clus1, background1, url ='pabies', type="go") {

  results <- list()
  
  print(clusterData1[clusterData1$cluster==gsub("i","", clus1),]$gene)
  eigen <- getEigenGenes(data1, clusterData1[clusterData1$cluster==gsub("i","", clus1),]$gene)
  
  genes1 <- if(grepl("i", clus1, fixed = TRUE)){
    print("negative")
    eigen$negative
  } else {
    eigen$positive
  }
  
  genes2 <- intersect(genes1, colnames(data2))

    p1 <- plotEigengene(data1, genes1, group1, time1)
    
    p2 <- plotEigengene(data2, genes2, group2, time2)
    
    enr1 <- gopher(genes2, background=background1, url=url, task = type)
    
    results$plot1 <- p1
    results$plot2 <- p2
    results$enr1 <- enr1
    
    
    return(results)
    
}
```
```{r}

spruceClus <- mostDegraded[7,]$`Spruce-control`
fungiClus <- mostDegraded[7,]$`Fungi-control`

spruceMulti <- enrichThis(controlSpruceData,  fertilisedSpruceData,
           controlSpruceClusters, 
           controlSpruceGroup, 
           controlSpruceTime, 
           fertilisedSpruceGroup, 
           fertilisedSpruceTime, 
           spruceClus, 
           controlSpruceClusters$gene, url="pabies", type=c("go","pfam","mapman","kegg"))

fungiMulti <- enrichThis(controlFungiData,  fertilisedFungiData,
           controlFungiClusters, 
           controlFungiGroup, 
           controlFungiTime, 
           fertilisedFungiGroup, 
           fertilisedFungiTime, 
           fungiClus, 
           controlFungiClusters$gene, url="ko", type="ko_pathway")

par(mfrow=c(2,2))
spruceMulti$plot1
spruceMulti$plot2
fungiMulti$plot1
fungiMulti$plot2
```






```{r}
#keep positive relations
condition.pos <- fertilised.res.sort[fertilised.res.sort$score>0,]
# condition.pos <- fertilised.res.sort
otherCondition.pos <-  control.res.sort[control.res.sort$score>0,]

conditionSpruceClusters <- fertilisedSpruceClusters
conditionFungiClusters <- fertilisedFungiClusters

data2Spruce <- controlSpruceData
data2Fungi  <- controlFungiData
data1Spruce <- fertilisedSpruceData
data1Fungi  <- fertilisedFungiData

group2Spruce <- controlSpruceGroup 
group2Fungi  <- controlFungiGroup 
group1Spruce <- fertilisedSpruceGroup 
group1Fungi  <- fertilisedFungiGroup 

time2Spruce <- controlSpruceTime
time2Fungi  <- controlFungiTime 
time1Spruce <- fertilisedSpruceTime
time1Fungi  <- fertilisedFungiTime 


# control2fertilised <- sapply(1:length(rownames(condition.pos)), function(l) {
score <- vector(mode = "numeric", length = length(rownames(condition.pos)))
scoreInControl <- vector(mode = "numeric", length = length(rownames(condition.pos)))

sapply(1:length(rownames(condition.pos)), function(l) {
  spruceCluster <- condition.pos[l,1]
  fungiCluster <- condition.pos[l,2]
  spruceReverse <- ifelse(endsWith(spruceCluster, "i"), TRUE, FALSE) 
  fungiReverse <- ifelse(endsWith(fungiCluster, "i"), TRUE, FALSE)

  spruceCluster <- ifelse(endsWith(spruceCluster, "i"), substr(spruceCluster, 1, nchar(spruceCluster)-1), spruceCluster) 
  fungiCluster <- ifelse(endsWith(fungiCluster, "i"), substr(fungiCluster, 1, nchar(fungiCluster)-1), fungiCluster) 

  c1SpruceClusterTotalGenes <- conditionSpruceClusters[conditionSpruceClusters$cluster==spruceCluster,]$gene
  c1FungiClusterTotalGenes <- conditionFungiClusters[conditionFungiClusters$cluster==fungiCluster,]$gene

  spruceGenes <- getEigenGenes(data1Spruce, c1SpruceClusterTotalGenes)
  c1SpruceClusterGenes <- if(spruceReverse) spruceGenes$negative else spruceGenes$positive

  fungiGenes <- getEigenGenes(data1Fungi, c1FungiClusterTotalGenes)
  c1FungiClusterGenes <- if(fungiReverse) fungiGenes$negative else fungiGenes$positive 

  # we extract only the genes that exist in both conditions and forget about uniques for the moment
  commonSpruceClusterGenes <- intersect(c1SpruceClusterGenes, colnames(data2Spruce))
  commonFungiClusterGenes <- intersect(c1FungiClusterGenes, colnames(data2Fungi))
  
  # TODO extract unique genes that exist in c1 but not in c2

  #initialise cor1 and cor2 for those cases that don't have genes in common
  cor1 <- 0
  cor2 <- 0
  if(length(commonSpruceClusterGenes)>0 && length(commonFungiClusterGenes)>0) {
    p1Spruce <- getPlotData(data1Spruce, commonSpruceClusterGenes, group1Spruce, time1Spruce)
    p1Fungi <- getPlotData(data1Fungi, commonFungiClusterGenes, group1Fungi, time1Fungi)
  
    p2Spruce <- getPlotData(data2Spruce, commonSpruceClusterGenes, group2Spruce, time2Spruce)
    p2Fungi <-getPlotData(data2Fungi, commonFungiClusterGenes, group2Fungi, time2Fungi)
    cor1 <- cor(p1Spruce$y, p1Fungi$y, method = c("spearman"))
    cor2 <- cor(p2Spruce$y, p2Fungi$y, method = c("spearman"))
  }

  score[l] <<- cor1
  scoreInControl[l] <<- cor2
  
})

fertilised2control <- cbind(condition.pos[,1:2], score, scoreInControl)
fertilised2control$degradation <- fertilised2control$score - fertilised2control$scoreInControl
fertilised2control.sort <- fertilised2control[order(fertilised2control$degradation, decreasing = F),]
#
fertilised2control.pos <- fertilised2control[fertilised2control$score>0,]
fertilised2control.pos.sort <- fertilised2control.pos[order(fertilised2control.pos$degradation, decreasing = F),]

write.table(fertilised2control.sort, file="fertilised2control.sort.tsv", quote = F, sep='\t', row.names = F)
```
```{r}
d <- melt(fertilised2control.sort[,c("score","scoreInControl")])
 
# Basic violin plot
p <- ggplot(d, aes(x=variable, y=value  )) + 
  geom_violin()
p

```

```{r eval=FALSE, include=FALSE}
# 
# plotEigengene(controlDataFungi,
#               controlClustersFungi[controlClustersFungi$cluster=="Cluster1",]$gene, 
#               fungiGroup,
#               fungiTime)
#               
# fungiProfile <- plotEigengene(controlDataFungi,
#               controlClustersFungi[controlClustersFungi$cluster=="Cluster1",]$gene, 
#               substr(rownames(controlDataFungi),4,4),
#               as.integer(substr(rownames(controlDataFungi),2,3))
#               )
# 
# fungiProfileData <- fungiProfile$data %>%
#   group_by(x) %>%
#   summarise(
#     n =n(),
#     y = mean(y, na.rm = TRUE)
#   )
# spruceProfile <- plotEigengene(controlDataSpruce,
#               controlClustersSpruce[controlClustersSpruce$cluster=="Cluster1",]$gene, 
#               substr(rownames(controlDataSpruce),4,4),
#               as.integer(substr(rownames(controlDataSpruce),2,3))
#               )
# 
# spruceProfileData <- spruceProfile$data %>%
#   group_by(x) %>%
#   summarise(
#     n =n(),
#     y = mean(y, na.rm = TRUE)
#   )

```
```{r eval=FALSE, include=FALSE}
# t <- controlDataSpruce[,
# controlClustersSpruce[controlClustersSpruce$cluster=="Cluster1",]$gene
# ]
# t.pca <- prcomp(t)
# pc1 <- t.pca$x[,1]
# pc1 <- scale(pc1)
# 
# sum(sign(cor(t.pca$x[,1,drop = FALSE], t)))
# t.pca$x[,1]
# 
# tx <- cor(t.pca$x[,1,drop = FALSE], t)
```

```{r}
save.image(file.path(here("env"),"env.RData"))
```


