---
title: "Flakaliden spruce roots + fungi cortinarius_glaucopus cluster analysis"
author: "Alonso Serrano"
date: "`r format(Sys.time(), '%d %B %Y')`"
output: 
  html_document: 
    number_sections: yes
    toc: yes
---

<style>
* {
  box-sizing: border-box;
}

/* Create two equal columns that floats next to each other */
.column {
  float: left;
  width: 50%;
  padding: 10px;
  
}

/* Clear floats after the columns */
.row:after {
  content: "";
  display: table;
  clear: both;
}
</style>

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```
# Introduction
This file contains the script and results of the data from spruce roots and fungi 
from the 2011 Flakaliden dataset. 

## Prerequisites
The variance stabilization transformed dataset, either as TSV or RData obtained from de differential expression analysis step.
Infomap clustering results from Seidr network

## Setup
```{r generalSetup, message=FALSE, warning=FALSE}
# preload packages
library(ggplot2)
library(RColorBrewer)
library(reshape2)
library(tidyr)
library(tibble)
library(jsonlite)
library(treemap)
library(KEGGREST)
library(reshape2)
library(data.table)

library(dplyr)
library(here)
source("~/Git/Rtoolbox/src/plotEigenGene.R")
source("~/Git/Rtoolbox/src/getEigengenes.R")
source("~/Git/Rtoolbox/src/plotEnrichedTreemap.R")
source("~/Git/UPSCb-common/src/R/gopher.R")
source("~/scripts/koPathwayDiseaesCleanup/koProcessing.R")
conditions <- c("control", "fertilised", "combined")

#specie <- "cenococcum_geophilum"
specie <- "cortinarius_glaucopus"
#specie <- "hyaloscypha_bicolor"
#specie <- "hyaloscypha_variabilis"
#specie <- "piloderma_croceum"


spruceDir <- "/mnt/picea/projects/spruce/vhurry/14_SpruceRoots_Project"
fungiDir <- file.path("/mnt/picea/projects/spruce/vhurry/fungi-flakaliden-2011/fungi-species",specie)

spruceDE <- file.path(spruceDir,"DE")
fungiDE <- file.path(fungiDir,"DE")

checkFile <- function(fileName) {
  if (!file.exists(fileName)) stop(paste("File", fileName, "doesn't exist"))
}
```

```{r, include = F}
library(stringi)
defChunkHook <- knitr::knit_hooks$get("chunk")
knitr::knit_hooks$set(chunk = function(x, options) {
  x <- defChunkHook(x, options)
  if(!is.null(options$multi.col)) {
    x    <- gsub("(\n\`\`\`\n##.*?\`\`\`)", "<div>\\1\n</div>", x)  # wrap div around output chunks
    x    <- gsub("(<img.*?>)", "<div>\\1\n</div>", x)               # wrap div around plots
    ncol <- nrow(stri_locate_all(x, regex = "<div.*?>")[[1]])       # get the number of div elements created
    x    <- gsub("<div>", paste0("<div style=\"width:", 100/ncol,"%;\">"), x)  # add the width to the divs
    x    <- paste0("<div class=\"multi-col\" style=\"display: flex; justify-content: center; align-items: center;\">\n", x, "</div>")  # wrap the mother div around all of the output
  }
  x
})
```
## Auxiliary functions

Get profile data
```{r miniFunction}
getPlotData <-function(data1, genes1, group1, time1) {
  p1 <- plotEigengene(data1, genes1, group1, time1)
    p1.data <- p1$data %>%
      group_by(x) %>%
      summarise(n = n(), y = mean(y, na.rm = TRUE) )
  return(p1.data)
}
```

Profile matching function
```{r profilefunction}
#TODO fix this using the minifunction
profileSimilarity <- function(data1, data2, clusterData1, clusterData2, 
                              group1, group2, time1, time2, name1, name2, minSize=20){

  #resultados
  # cluster1F -> cluster1S -> 0.34
  clusters1 <- unique(clusterData1$cluster)
  clusters2 <- unique(clusterData2$cluster)
  
  res <- data.frame(name1 = character(),
                    genes1 = character(),
                   name2 = character(), 
                   genes2 = character(),
                   score = double(), 
                   stringsAsFactors = FALSE) 

  #por cada cluster
  sapply(clusters1, function(clus1){
    
    clus1Genes <- clusterData1[clusterData1$cluster==clus1,]$gene
    clus1Genes_split <- getEigenGenes(data1, clus1Genes)
    clus1i <- paste0(clus1, "i")
    lenClus1 <- length(clus1Genes_split$positive)
    lenClus1i <- length(clus1Genes_split$negative)
    
    p1 <- plotEigengene(data1, clus1Genes, group1, time1)
    p1.data <- p1$data %>%
      group_by(x) %>%
      summarise(n = n(), y = mean(y, na.rm = TRUE) )
    eigen1 <- p1.data$y
    eigen1i <- p1.data$y * -1
  
    sapply(clusters2, function(clus2){
      #print(paste("comparing",clus1, clus2))
      
      clus2Genes <- clusterData2[clusterData2$cluster==clus2,]$gene
      clus2Genes_split <- getEigenGenes(data2, clus2Genes)
      clus2i <- paste0(clus2, "i")
      lenClus2 <- length(clus2Genes_split$positive)
      lenClus2i <- length(clus2Genes_split$negative)
    
        p2 <- plotEigengene(data2,clus2Genes, group2, time2)
        p2.data <- p2$data %>%
          group_by(x) %>%
          summarise(n = n(), y = mean(y, na.rm = TRUE) )
        eigen2 <- p2.data$y
        eigen2i <- p2.data$y * -1
        
        #add comparitions
        #1vs2
        if((lenClus1 > minSize) && (lenClus2 > minSize))
        res <<- rbind(res, c(clus1, lenClus1, clus2, lenClus2, 
                             cor(eigen1, eigen2, method = c("spearman"))), stringsAsFactors = FALSE)
        #1ivs2
        if((lenClus1i > minSize) && (lenClus2 > minSize))
        res <<- rbind(res, c(clus1i, lenClus1i, clus2, lenClus2, 
                             cor(eigen1i, eigen2, method = c("spearman"))), stringsAsFactors = FALSE)
        #1vs2i
        if((lenClus1 > minSize) && (lenClus2i > minSize))
        res <<- rbind(res, c(clus1, lenClus1, clus2i, lenClus2i,
                             cor(eigen1, eigen2i, method = c("spearman"))), stringsAsFactors = FALSE)
        #1ivs2i
        if((lenClus1i > minSize) && (lenClus2i > minSize))
        res <<- rbind(res, c(clus1i, lenClus1i, clus2i, lenClus2i, 
                             cor(eigen1i, eigen2i, method = c("spearman"))), stringsAsFactors = FALSE)
    })

  })
  # retorno
  names(res) <- c(name1, paste0(name1,"-genes"),name2, paste0(name2,"-genes"), "score")
  res$score <- as.numeric(res$score)
  return(res)
}
```

Turn clusters into networks
```{r getClustersAmounts, include=F, eval=F}
# future uses
cluster2Network <- function(data, clusterData, species){
  
  temp <- lapply(unique(clusterData$cluster), function(localCluster) {
    spruceGenes <- getEigenGenes(data, clusterData[clusterData$cluster==localCluster,]$gene)
  })
  #spruceCluster <- "Cluster1"
  names(temp) <- paste0(species, unique(clusterData$cluster) )
      
  posTemp <- lapply(temp, function(x){
    x$positive
  })
  
  negTemp <- lapply(temp, function(x){
    x$negative
  })
  names(negTemp) <- paste0(names(negTemp),"i")
  
  posResult <- sapply(posTemp, function(x) {
    length(x)
  })
  negResult <- sapply(negTemp, function(x) {
    length(x)
  })
  
  posDF <- data.frame(clusters=names(posResult), geneNumber=posResult)
  negDF <- data.frame(clusters=names(negResult), geneNumber=negResult)
  res <- rbind(posDF, negDF)
  return (res)
}


```

Create a profile degradation matrix
```{r profileDegradationFunction}

getProfileDegradationMatrix <- function(pairedData, conditionSpruceClusters, conditionFungiClusters, 
                                  data1Spruce, data1Fungi,data2Spruce, data2Fungi,
                                  group1Spruce, group1Fungi,group2Spruce, group2Fungi,
                                  time1Spruce, time1Fungi,time2Spruce, time2Fungi,  
                                  
                                  conditionNames = c("control", "fertilised") ){

  condition.pos <- pairedData #[pairedData$score>0,]
  
  score1 <- vector(mode = "numeric", length = length(rownames(condition.pos)))
  score2 <- vector(mode = "numeric", length = length(rownames(condition.pos)))
  
  sapply(1:length(rownames(condition.pos)), function(l) {
    
   
    spruceCluster <- condition.pos[l,1]
    fungiCluster <- condition.pos[l,3]
    spruceReverse <- ifelse(endsWith(spruceCluster, "i"), TRUE, FALSE) 
    fungiReverse <- ifelse(endsWith(fungiCluster, "i"), TRUE, FALSE)
  
    spruceCluster <- ifelse(endsWith(spruceCluster, "i"), substr(spruceCluster, 1, nchar(spruceCluster)-1), spruceCluster) 
    fungiCluster <- ifelse(endsWith(fungiCluster, "i"), substr(fungiCluster, 1, nchar(fungiCluster)-1), fungiCluster) 
  
    c1SpruceClusterTotalGenes <- conditionSpruceClusters[conditionSpruceClusters$cluster==spruceCluster,]$gene
    c1FungiClusterTotalGenes <- conditionFungiClusters[conditionFungiClusters$cluster==fungiCluster,]$gene
  
    spruceGenes <- getEigenGenes(data1Spruce, c1SpruceClusterTotalGenes)
    c1SpruceClusterGenes <- if(spruceReverse) spruceGenes$negative else spruceGenes$positive
  
    fungiGenes <- getEigenGenes(data1Fungi, c1FungiClusterTotalGenes)
    c1FungiClusterGenes <- if(fungiReverse) fungiGenes$negative else fungiGenes$positive 
  
    # we extract only the genes that exist in both conditions and forget about uniques for the moment
    commonSpruceClusterGenes <- intersect(c1SpruceClusterGenes, colnames(data2Spruce))
    commonFungiClusterGenes <- intersect(c1FungiClusterGenes, colnames(data2Fungi))
    
    # TODO extract unique genes that exist in c1 but not in c2
  
    #initialise cor1 and cor2 for those cases that don't have genes in common
    cor1 <- 0
    cor2 <- 0
    if(length(commonSpruceClusterGenes)>0 && length(commonFungiClusterGenes)>0) {
      p1Spruce <- getPlotData(data1Spruce, commonSpruceClusterGenes, group1Spruce, time1Spruce)
      p1Fungi <- getPlotData(data1Fungi, commonFungiClusterGenes, group1Fungi, time1Fungi)
    
      p2Spruce <- getPlotData(data2Spruce, commonSpruceClusterGenes, group2Spruce, time2Spruce)
      p2Fungi <-getPlotData(data2Fungi, commonFungiClusterGenes, group2Fungi, time2Fungi)
      cor1 <- cor(p1Spruce$y, p1Fungi$y, method = c("spearman"))
      cor2 <- cor(p2Spruce$y, p2Fungi$y, method = c("spearman"))
    }
  
    score1[l] <<- cor1
    score2[l] <<- cor2
    
  }) #end of sapply

  res <- condition.pos[,1:4]
  res[[paste0(conditionNames[1],"_score")]] <- score1
  res[[paste0(conditionNames[2],"_score")]] <- score2
  res$degradation <- score1 - score2
  res.sort <- res[order(res$degradation, decreasing = T),]
  
  #write.table(fertilised2control.sort, file="fertilised2control.sort.tsv", quote = F, sep='\t', row.names = F)
  #
  res.sort

}
```


## Check files

### VSD data file
The files are tab separated, rows contain sample meta information (week and treatment),
columns contain gene or KEGG orthologs names. The values are expression data after variance stabilization transformation

### Cluster files
They must contain two columns gene and cluster and be tab separated<br />
ex.<br />
gene	cluster<br />
MA_5320g0010	Cluster1<br />
MA_32763g0010	Cluster1<br />
MA_66378g0010	Cluster1<br />


```{r loadRdata}
load(file.path(spruceDE, "controlVsdData.RData"))
controlSpruceData <- controlData
load(file.path(spruceDE, "fertilisedVsdData.RData"))
fertilisedSpruceData <- fertilisedData
# load(file.path(spruceDE, "combinedVsdData.RData"))
# combinedSpruceData <- combinedData

load(file.path(fungiDE, "controlVsdData.RData"))
controlFungiData <- controlData
load(file.path(fungiDE, "fertilisedVsdData.RData"))
fertilisedFungiData <- fertilisedData


```

```{r}
# load spruce data files
controlSpruceClusterFile <- file.path(spruceDir,"spruce-networks/control/cluster","InfomapClusters.tsv")
checkFile(controlSpruceClusterFile)

# load fungi data files
controlFungiClusterFile <- file.path(fungiDir,"networks/control/results/aggregated","infomapClusters.tsv")
checkFile(controlFungiClusterFile)
```

## Load files
```{r loadFiles}
#name correction, spruce rownames, have an extra dash
rownames(controlSpruceData) <- gsub("-", "", rownames(controlSpruceData))
rownames(fertilisedSpruceData) <- gsub("-", "", rownames(fertilisedSpruceData))

controlSpruceClusters <- read.table(controlSpruceClusterFile, stringsAsFactors=F, header=T, sep='\t', comment.char="")
controlFungiClusters <- read.table(controlFungiClusterFile, stringsAsFactors=F, header=T, sep='\t', comment.char="")
```

## Create metadata based on the information on sample names
```{r}
controlSpruceGroup <- substr(rownames(controlSpruceData),4,4)
controlSpruceTime <- as.integer(substr(rownames(controlSpruceData),2,3))

controlFungiGroup <- substr(rownames(controlFungiData),4,4)
controlFungiTime <- as.integer(substr(rownames(controlFungiData),2,3))

fertilisedSpruceGroup <- substr(rownames(fertilisedSpruceData),4,4)
fertilisedSpruceTime <- as.integer(substr(rownames(fertilisedSpruceData),2,3))

fertilisedFungiGroup <- substr(rownames(fertilisedFungiData),4,4)
fertilisedFungiTime <- as.integer(substr(rownames(fertilisedFungiData),2,3))
```

# Control profile matching
```{r controlMatching, echo=TRUE, message=FALSE}
control.res <- profileSimilarity(controlSpruceData, controlFungiData, controlSpruceClusters, controlFungiClusters, controlSpruceGroup, controlFungiGroup, controlSpruceTime, controlFungiTime, 
                         "Spruce-control", "Fungi-control")

control.res.sort <- control.res[order(control.res$score, decreasing = T),]
```

## Best correlated
```{r controlHead}
head(control.res.sort)
```
## Best reverse correlated
```{r controlTail}
tail(control.res.sort)
```
## Save results to file
```{r controlSave}
# control.res.sort$`Spruce-control`<- paste0("spruce",control.res.sort$`Spruce-control`)
# control.res.sort$`Fungi-control`<- paste0("fungi",control.res.sort$`Fungi-control`)
write.table(control.res.sort, file=here("results",paste0("controlProfileMatching_",specie,"_",Sys.Date(),".tsv")), sep='\t', quote=F, row.names=F)
```

```{r eval=FALSE, include=FALSE}
spruceControl <- cluster2Network(controlSpruceData, controlSpruceClusters, "spruce")
fungiControl <- cluster2Network(controlFungiData, controlFungiClusters, "fungi")
write.table(rbind(spruceControl, fungiControl), 
            file=here("results",paste0("controlClusterNetworkMeta",specie,"_",Sys.Date(),".tsv")),
             sep='\t', row.names = F, quote=F)

# spruceFertilised <- cluster2Network(fertilisedSpruceData, fertilisedSpruceClusters, "spruce")
# fungiFertilised <- cluster2Network(fertilisedFungiData, fertilisedFungiClusters, "fungi")
# 
# write.table(rbind(spruceFertilised, fungiFertilised),
#             file=here("results",paste0("fertilisedClusterNetworkMeta",specie,"_",Sys.Date(),".tsv"))
#             , sep='\t', row.names = F, quote=F)
```

## Obtain profile degradation matrix
```{r}
load(here("env/cortinarius_controlDEG.RData"))
```


```{r controlDEF, eval=F}
controlDEG <- getProfileDegradationMatrix(control.res.sort, 
                                          conditionSpruceClusters=controlSpruceClusters, 
                                          conditionFungiClusters=controlFungiClusters,
                                          data1Spruce=controlSpruceData, 
                                          data1Fungi=controlFungiData, 
                                          data2Spruce=fertilisedSpruceData, 
                                          data2Fungi=fertilisedFungiData,
                                          group1Spruce=controlSpruceGroup, 
                                          group1Fungi=controlFungiGroup,
                                          group2Spruce=fertilisedSpruceGroup , 
                                          group2Fungi=fertilisedFungiGroup ,
                                          time1Spruce=controlSpruceTime , 
                                          time1Fungi=controlFungiTime,
                                          time2Spruce=fertilisedSpruceTime, 
                                          time2Fungi=fertilisedFungiTime ,
                                          conditionNames = c("control", "fertilised")
                                          
                                          )


```

```{r}
write.table(controlDEG, file=here("results", paste0("controlDEGfull_",specie,"_",Sys.Date(),".tsv")), quote = F, sep='\t', row.names = F)
```

```{r, eval=F}
save(controlDEG, file=here("env/cortinarius_controlDEG.RData"))
```

## Reduced violin plot
```{r controlViolin, warning=FALSE}
controlDEG.df <- controlDEG[controlDEG$control_score>0,]
controlDEG.df <- controlDEG.df[order(controlDEG.df$degradation, decreasing = F),]
controlDEG.df <- melt(controlDEG.df[,c("control_score","fertilised_score")])
 
p <- ggplot(controlDEG.df, aes(x=variable, y=value  )) + 
  geom_violin() +
  stat_summary(fun.y=mean, geom="point", shape=23, size=2) + 
  stat_summary(fun.y=median, geom="point", size=2, color="red") +
  geom_boxplot(width=0.1)

print(p)
```

```{r eval=FALSE, include=FALSE}
png(here("results", paste0("controlDEGplot_",specie,"_",Sys.Date(),".png")),
 units="px", width=1920, height=1080, res=300)
print(p)
dev.off()

write.table(controlDEG.df, file=here("results", paste0("controlDEG_",specie,"_",Sys.Date(),"_df_sort.tsv")), quote = F, sep='\t', row.names = F)

```

## Full violin plot
```{r controlViolin2, warning=FALSE}
controlDEG.df <- controlDEG
controlDEG.df <- controlDEG.df[order(controlDEG.df$degradation, decreasing = F),]
controlDEG.df <- melt(controlDEG.df[,c("control_score","fertilised_score")])
 
p <- ggplot(controlDEG.df, aes(x=variable, y=value  )) + 
  geom_violin() +
  stat_summary(fun.y=mean, geom="point", shape=23, size=2) + 
  stat_summary(fun.y=median, geom="point", size=2, color="red") +
  geom_boxplot(width=0.1)

print(p)
```

```{r eval=FALSE, include=FALSE}
 # png(here("results", paste0("controlDEGplot_",specie,"_",Sys.Date(),"_full.png")),
   #units="px", width=1920, height=1080, res=300)
  print(p)
 # dev.off()
```


```{r}
controlAvg <- mean(controlDEG$degradation)
print(controlAvg)
```

## Significant degradations
Get the maximum difference

```{r}
res.pos <- controlDEG %>% 
  filter((control_score > 0) & (fertilised_score > 0)) #%>% 
  #filter((degradation<1) & (degradation > -1))


res.pos <- res.pos[order(res.pos$degradation, decreasing = T),]

res.neg <- controlDEG %>% 
  filter((control_score < 0) & (fertilised_score < 0)) #%>% 
  #filter((degradation<1) & (degradation > -1))


res.neg <- res.neg[order(res.neg$degradation, decreasing = T),]

write.table(rbind(res.pos, res.neg), file=here("results", paste0("maxDifference_",specie,"_",Sys.Date(),".tsv")), quote = F, sep='\t', row.names = F)
```


Get the top 20 most important changes(10 degradations, 10 recovers)
```{r}
mostDegraded <- rbind(head(res.pos, n=5),
                      tail(res.pos, n=5),
                      head(res.neg, n=5),
                      tail(res.neg, n=5))
mostDegraded
```

Enrich & plot results
```{r}
enrichThis <- function(data1, data2, clusterData1, clus1, background1, url ='pabies', type="go") {

  eigen <- getEigenGenes(data1, clusterData1[clusterData1$cluster==gsub("i","", clus1),]$gene)
  
  genes1 <- if(grepl("i", clus1, fixed = TRUE)){
    eigen$negative
  } else {
    eigen$positive
  }
    
  genes2 <- intersect(genes1, colnames(data2))
  genes3 <- intersect(genes1, genes2)
  
  if(length(genes3)>1) {
    enr1 <- gopher(genes3, background=background1, url=url, task = type)
  } else {
    enr1 <- NULL
  }
  return(enr1)
}
```

```{r, eval=T, include=F}
load(file=here("env/cortinarius_enrRes.RData"))
```

```{r superLoop, eval=F}
enrRes <- lapply(1:20, function(x) {
  spruceClus <- mostDegraded[x,]$`Spruce-control`
  fungiClus <- mostDegraded[x,]$`Fungi-control`

  res <- list()
  spruceMulti <- enrichThis(controlSpruceData,  fertilisedSpruceData,
             controlSpruceClusters, spruceClus,
             controlSpruceClusters$gene, url="pabies", type=c("go","pfam","mapman","kegg")
             )
  # only work with the top 5 species
  fungiMulti <- enrichThis(controlFungiData,  fertilisedFungiData,
             controlFungiClusters, fungiClus,
             controlFungiClusters$gene, url=paste0("fungi2011_",specie), type="go"
             )

  fungiMultiGeneric <- enrichThis(controlFungiData,  fertilisedFungiData,
           controlFungiClusters, fungiClus,
           controlFungiClusters$gene, url="fungi2011", type=c("cog","ko","ko_pathway","kog") 
           )
  
  res$spruce <- spruceMulti
  res$fungi <- c(fungiMulti, fungiMultiGeneric)
  res
})
```

```{r, eval=F,include=F}
save(enrRes, file=here("env/cortinarius_enrRes.RData"))
```


```{r}
plotThis <- function(data1, data2, clusterData1, group1, time1, group2, time2, clus1, title1="", title2="") {

  results <- list()
  eigen <- getEigenGenes(data1, clusterData1[clusterData1$cluster==gsub("i","", clus1),]$gene)
  
  genes1 <- if(grepl("i", clus1, fixed = TRUE)){
    eigen$negative
  }else {
    eigen$positive
    }
    
  genes2 <- intersect(genes1, colnames(data2))

  p1 <- plotEigengene(data1, genes1, group1, time1, title=title1)
  p2 <- plotEigengene(data2, genes2, group2, time2, title=title2, colors="#BE9230")
  genes3 <- intersect(genes1, genes2)
  #data3 <- rbind(data1[,genes3], data2[,genes3])
  
  # p3 <- plotEigengene(data3, genes3, c(group1, group2),
  #                     c(time1,time2), title="", noLegend = F)

   if(length(genes3)>0) {
      plotData1 <- getPlotData(data1, genes3, group1, time1)
      plotData2 <- getPlotData(data2, genes3, group2, time2)
      cor1 <- cor(plotData1$y, plotData2$y, method = c("spearman"))
  } #else { 0 }
  
  results$plot1 <- p1
  results$plot2 <- p2
  results$cor1 <- cor1
  #results$plot3 <- p3

  return(results)
}
```


## Enrich & plot
```{r superLoop2, results='asis', fig.hold='hold', out.width="50%"}
None <- lapply(1:20, function(x) {
  spruceClus <- mostDegraded[x,]$`Spruce-control`
  fungiClus <- mostDegraded[x,]$`Fungi-control`
  cat("\n")
  cat("<hr>")
  cat(paste0("<H3>Spruce cluster: ", spruceClus, ". Fungi cluster ",fungiClus, ". Degradation: ", round(mostDegraded[x,]$degradation, 4),"</H3>"))
  cat("\n\n")
  spruceMulti <- plotThis(controlSpruceData,  fertilisedSpruceData,
             controlSpruceClusters,              controlSpruceGroup, 
             controlSpruceTime, 
             fertilisedSpruceGroup, 
             fertilisedSpruceTime, 
             spruceClus, 
            
             title1 = paste("Spruce control cluster", spruceClus, "with control data"),
             title2 = paste("Spruce control cluster", spruceClus, "with fertilised data")
            
             )
  
  fungiMulti <- plotThis(controlFungiData,  fertilisedFungiData,
             controlFungiClusters, 
             controlFungiGroup, 
             controlFungiTime, 
             fertilisedFungiGroup, 
             fertilisedFungiTime, 
             fungiClus, 

             title1 = paste("Fungi control cluster", fungiClus, "with control data"),
             title2 = paste("Fungi control cluster", fungiClus, "with fertilised data")
             )

  cat(paste("<H4>Profiles plot</H4>"))
  cat("\n\n")
  cat(paste("<center>Spruce - fungi control profiles correlation",round(mostDegraded[x,]$control_score,4), "</center>"))
  cat("\n\n")
  par(mar = c(4, 4, 0.1, 0.1))
  print(spruceMulti$plot1)
  print(fungiMulti$plot1)
  cat(paste("<div class='row'>",
    "<div class='column'><center>Spruce correlation", round(spruceMulti$cor1, 4), "</center></div>",
    "<div class='column'><center>Fungi correlation", round(fungiMulti$cor1, 4), "</center></div></div>"
  ))

  print(spruceMulti$plot2)
  print(fungiMulti$plot2)
  par(mfrow=c(1,1))
  cat("\n\n")
  cat("\n\n")
  cat(paste("<center>Spruce - fungi fertilised profiles correlation",round(mostDegraded[x,]$fertilised_score,4),"</center>"))
  cat("\n\n")
  cat("\n\n")

  cat(paste("<H4>Spruce treemaps</H4>"))
  cat("\n\n")
  par(mar = c(4, 4, 0.1, 0.1))
  None <- sapply(c("go","pfam","mapman","kegg"), function(y){
    if(!is.null(enrRes[[x]]$spruce[[y]])) {
       if(y=="kegg"){
         enrRes[[x]]$spruce[[y]] <- koTranslate(enrRes[[x]]$spruce[[y]])
       }
      plotEnrichedTreemap(enrRes[[x]]$spruce, enrichment = y, namespace = "none", title=paste(y,"enrichment"))
    }
  })
  par(mfrow=c(1,1))
  cat("\n\n")
  cat("<hr>")

  cat(paste("<H4>Fungi treemaps</H4>"))
  cat("\n\n")
  par(mar = c(4, 4, 0.1, 0.1))
  
  None <- sapply(c("go", "cog","ko","ko_pathway","kog"), function(y){
    if(!is.null(enrRes[[x]]$fungi[[y]])) {
      if(y=="ko" || y=="ko_pathway"){
        enrRes[[x]]$fungi[[y]] <- koTranslate(enrRes[[x]]$fungi[[y]])
      }
      plotEnrichedTreemap(enrRes[[x]]$fungi, enrichment = y, namespace = "none",title=paste(y,"enrichment"))
    }

  })
  par(mfrow=c(1,1))
  cat("\n\n")
  cat("<hr>")
  cat("<hr>")
})

```


```{r, eval=F, include=F}
save.image(file.path(here("env"),"env.RData"))
```

# Session information
```{r}
sessionInfo()
```
